<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Inferencia Bayesiana - 6&nbsp; Estimación de Modelos Bayesianos (Parte 2)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./estimacion_modelos1.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./estimacion_modelos2.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Estimación de Modelos Bayesianos (Parte 2)</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Inferencia Bayesiana</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prefacio</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pensamiento.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Pensamiento Bayesiano</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./unparametro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Modelos de un parámetro</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./variosparametros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Modelos de varios parámetros</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./computacion_bayes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Computación Bayesiana</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./estimacion_modelos1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Estimación de Modelos Bayesianos (Parte 1)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./estimacion_modelos2.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Estimación de Modelos Bayesianos (Parte 2)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#comparación-de-medias" id="toc-comparación-de-medias" class="nav-link active" data-scroll-target="#comparación-de-medias"><span class="header-section-number">6.1</span> Comparación de medias</a>
  <ul class="collapse">
  <li><a href="#datos" id="toc-datos" class="nav-link" data-scroll-target="#datos"><span class="header-section-number">6.1.1</span> Datos</a></li>
  <li><a href="#modelo" id="toc-modelo" class="nav-link" data-scroll-target="#modelo"><span class="header-section-number">6.1.2</span> Modelo</a></li>
  <li><a href="#desviaciones-estándar-distintas-para-cada-grupo" id="toc-desviaciones-estándar-distintas-para-cada-grupo" class="nav-link" data-scroll-target="#desviaciones-estándar-distintas-para-cada-grupo"><span class="header-section-number">6.1.3</span> Desviaciones estándar distintas para cada grupo:</a></li>
  <li><a href="#rope-región-de-equivalencia-práctica" id="toc-rope-región-de-equivalencia-práctica" class="nav-link" data-scroll-target="#rope-región-de-equivalencia-práctica"><span class="header-section-number">6.1.4</span> ROPE (Región de Equivalencia Práctica)</a></li>
  <li><a href="#comparaciones-pareadas" id="toc-comparaciones-pareadas" class="nav-link" data-scroll-target="#comparaciones-pareadas"><span class="header-section-number">6.1.5</span> Comparaciones Pareadas</a></li>
  </ul></li>
  <li><a href="#regresión-lineal-simple" id="toc-regresión-lineal-simple" class="nav-link" data-scroll-target="#regresión-lineal-simple"><span class="header-section-number">6.2</span> Regresión Lineal Simple</a>
  <ul class="collapse">
  <li><a href="#ejemplo-datos-de-galton" id="toc-ejemplo-datos-de-galton" class="nav-link" data-scroll-target="#ejemplo-datos-de-galton"><span class="header-section-number">6.2.1</span> Ejemplo: Datos de Galton</a></li>
  <li><a href="#modelo-de-regresión" id="toc-modelo-de-regresión" class="nav-link" data-scroll-target="#modelo-de-regresión"><span class="header-section-number">6.2.2</span> Modelo de regresión</a></li>
  <li><a href="#análisis-posterior" id="toc-análisis-posterior" class="nav-link" data-scroll-target="#análisis-posterior"><span class="header-section-number">6.2.3</span> Análisis posterior</a></li>
  <li><a href="#ajuste-de-modelos-con-stan" id="toc-ajuste-de-modelos-con-stan" class="nav-link" data-scroll-target="#ajuste-de-modelos-con-stan"><span class="header-section-number">6.2.4</span> Ajuste de modelos con Stan</a></li>
  <li><a href="#modelo-con-dos-interceptos" id="toc-modelo-con-dos-interceptos" class="nav-link" data-scroll-target="#modelo-con-dos-interceptos"><span class="header-section-number">6.2.5</span> Modelo con dos interceptos</a></li>
  </ul></li>
  <li><a href="#regresión-lineal-múltiple" id="toc-regresión-lineal-múltiple" class="nav-link" data-scroll-target="#regresión-lineal-múltiple"><span class="header-section-number">6.3</span> Regresión Lineal Múltiple</a>
  <ul class="collapse">
  <li><a href="#ejemplo-sat" id="toc-ejemplo-sat" class="nav-link" data-scroll-target="#ejemplo-sat"><span class="header-section-number">6.3.1</span> Ejemplo: SAT</a></li>
  </ul></li>
  <li><a href="#ajuste-de-un-modelo-lineal-con-brms" id="toc-ajuste-de-un-modelo-lineal-con-brms" class="nav-link" data-scroll-target="#ajuste-de-un-modelo-lineal-con-brms"><span class="header-section-number">6.4</span> Ajuste de un modelo lineal con brms</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Estimación de Modelos Bayesianos (Parte 2)</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="comparación-de-medias" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="comparación-de-medias"><span class="header-section-number">6.1</span> Comparación de medias</h2>
<p>Cushny, A. R. and Peebles, A. R. (1905) “The action of optical isomers: II hyoscines.” <em>The Journal of Physiology</em> 32, 501–510.</p>
<section id="datos" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="datos"><span class="header-section-number">6.1.1</span> Datos</h3>
<p>Vamos a ver los datos. (<code>extra</code> = sueño adicional con el medicamento; <code>group</code> debería ser realmente <code>drug</code>, así que vamos a renombrarlo.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.0     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggformula)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: scales

Attaching package: 'scales'

The following object is masked from 'package:purrr':

    discard

The following object is masked from 'package:readr':

    col_factor

Loading required package: ggridges

New to ggformula?  Try the tutorials: 
    learnr::run_tutorial("introduction", package = "ggformula")
    learnr::run_tutorial("refining", package = "ggformula")</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>sleep <span class="ot">&lt;-</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  datasets<span class="sc">::</span>sleep <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">drug =</span> group)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_boxplot</span>(extra <span class="sc">~</span> drug, <span class="at">data =</span> sleep)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a> <span class="fu">df_stats</span>(extra <span class="sc">~</span> drug, <span class="at">data =</span> sleep)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  response drug  min     Q1 median   Q3 max mean       sd  n missing
1    extra    1 -1.6 -0.175   0.35 1.70 3.7 0.75 1.789010 10       0
2    extra    2 -0.1  0.875   1.75 4.15 5.5 2.33 2.002249 10       0</code></pre>
</div>
</div>
<p><strong>Problema de estudio</strong>: Se compararon los hábitos de sueño de los sujetos sin un medicamento para inducir el sueño y luego con él para ver cómo dos medicamentos diferentes afectaban de manera significativa el sueño.</p>
</section>
<section id="modelo" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="modelo"><span class="header-section-number">6.1.2</span> Modelo</h3>
<p>El punto de partida tradicional para modelar las medias es asumir que cada grupo se muestrea de una distribución normal con media desconocida y una <strong>desviación estándar común</strong>. (Veremos que no es más difícil tener desviaciones estándar diferentes).</p>
<p>Entonces, para dos grupos, nuestro modelo tiene tres parámetros: dos medias ((_1) y (_2)) y una desviación estándar ().</p>
<p>Por supuesto, también necesitamos previas para estos parámetros. Una previa común para las medias es una distribución normal. Comenzaremos con un previa uniforme para la desviación estándar, pero discutiremos mejores alternativas en breve.</p>
<p>Esto nos da el siguiente esquema para nuestro modelo:</p>
<p>[ <span class="math display">\[\begin{align*}
Y_{i|g} &amp; \sim \text{Norm}(\mu_g, \sigma) \\
\mu_g &amp; \sim \text{Norm}(?, ?) \\
\sigma &amp; \sim \text{Unif}(?, ?)
\end{align*}\]</span> ]</p>
<p>Los signos de interrogación se llenarán en función de consideraciones de la escala (orden de magnitud de los datos) y la cantidad de regularización que queremos hacer.</p>
<p>Algunas alternativas:</p>
<ul>
<li>Media para la previa en (_g): 0
<ul>
<li>Corresponde a que el medicamento no tenga impacto en el sueño</li>
<li>Permite que el medicamento aumente o disminuya el sueño sin prejuicios</li>
<li>Cualquier otro número requeriría más justificación.</li>
</ul></li>
<li>Desviación estándar para la previa en (_g): 3
<ul>
<li>Indica que estamos 95% seguros de que el impacto de un medicamento estará entre -6 y 6 horas adicionales de sueño y que es muy poco probable que el medicamento cambie el sueño en 9 o más horas. Esta es una previa bastante débil (6 horas adicionales de sueño sería mucho).</li>
</ul></li>
</ul>
<p>Planteamiento en JAGS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(R2jags)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: rjags</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: coda</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Linked to JAGS 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loaded modules: basemod,bugs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'R2jags'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:coda':

    traceplot</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sleep_model <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>Nobs) {</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    extra[i] <span class="sc">~</span> <span class="fu">dnorm</span>(mu[drug[i]], <span class="dv">1</span> <span class="sc">/</span> sigma<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (d <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>Ndrugs) {</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    mu[d] <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span><span class="sc">^</span><span class="dv">2</span>)          <span class="co"># sd = 3</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  sigma <span class="sc">~</span> <span class="fu">dunif</span>(<span class="dv">2</span><span class="sc">/</span><span class="dv">1000</span>, <span class="dv">2</span> <span class="sc">*</span> <span class="dv">1000</span>)    </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  delta_mu    <span class="ot">&lt;-</span> mu[<span class="dv">2</span>] <span class="sc">-</span> mu[<span class="dv">1</span>]</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  tau         <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> sigma<span class="sc">^</span><span class="dv">2</span>               </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>sleep_jags <span class="ot">&lt;-</span> </span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sleep_model,</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"mu"</span>, <span class="st">"sigma"</span>, <span class="st">"delta_mu"</span>),</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">extra =</span> sleep<span class="sc">$</span>extra,</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">drug  =</span> sleep<span class="sc">$</span>drug,</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">Nobs  =</span> <span class="fu">nrow</span>(sleep),</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">Ndrugs =</span> <span class="dv">2</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">DIC =</span> <span class="cn">FALSE</span>  <span class="co"># because we haven't discussed deviance yet</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>module glm loaded</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>module dic loaded</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Compiling model graph
   Resolving undeclared variables
   Allocating nodes
Graph information:
   Observed stochastic nodes: 20
   Unobserved stochastic nodes: 3
   Total graph size: 57

Initializing model</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(CalvinBayes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: bayesplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is bayesplot version 1.11.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- Online documentation and vignettes at mc-stan.org/bayesplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- bayesplot theme set to bayesplot::theme_default()</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>   * Does _not_ affect other ggplot2 plots</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>   * See ?bayesplot_theme_set for details on theme setting</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'CalvinBayes'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:bayesplot':

    rhat</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:datasets':

    HairEyeColor</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bayesplot)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sleep_jags)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fit using jags
 3 chains, each with 2000 iterations (first 1000 discarded)
 n.sims = 3000 iterations saved
         mu.vect sd.vect   2.5%   25%   50%   75% 97.5%  Rhat n.eff
delta_mu   1.514   0.896 -0.286 0.934 1.547 2.101 3.242 1.002  3000
mu[1]      0.719   0.632 -0.531 0.286 0.724 1.136 1.930 1.001  3000
mu[2]      2.233   0.648  0.923 1.824 2.228 2.649 3.480 1.001  3000
sigma      2.039   0.365  1.454 1.775 1.996 2.258 2.885 1.003   850</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sleep_mcmc <span class="ot">&lt;-</span> <span class="fu">as.mcmc</span>(sleep_jags)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">regex_pars =</span> <span class="st">"mu"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">regex_pars =</span> <span class="st">"sigma"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_pairs</span>(sleep_mcmc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Cuantificamos la probabilidad posterior de que el tiempo medio de sueño fue superior en la segunda droga que en la primera:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mosaic)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'mosaic':
  method                           from   
  fortify.SpatialPolygonsDataFrame ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
The 'mosaic' package masks several functions from core packages in order to add 
additional features.  The original behavior of these functions should not be affected by this.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'mosaic'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Matrix':

    mean</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:scales':

    rescale</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    count, do, tally</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    cross</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ggplot2':

    stat</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    binom.test, cor, cor.test, cov, fivenum, IQR, median, prop.test,
    quantile, sd, t.test, var</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    max, mean, min, prod, range, sample, sum</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prop</span>( <span class="sc">~</span>(delta_mu <span class="sc">&gt;</span> <span class="dv">0</span>), <span class="at">data =</span> <span class="fu">posterior</span>(sleep_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>prop_TRUE 
0.9523333 </code></pre>
</div>
</div>
</section>
<section id="desviaciones-estándar-distintas-para-cada-grupo" class="level3" data-number="6.1.3">
<h3 data-number="6.1.3" class="anchored" data-anchor-id="desviaciones-estándar-distintas-para-cada-grupo"><span class="header-section-number">6.1.3</span> Desviaciones estándar distintas para cada grupo:</h3>
<p>Si consideramos una desviación estándar distinta para cada grupo:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>sleep_model2 <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>Nobs) {</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    extra[i] <span class="sc">~</span> <span class="fu">dnorm</span>(mu[drug[i]], <span class="dv">1</span><span class="sc">/</span>sigma[drug[i]]<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (d <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>Ndrugs) {</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>    mu[d]    <span class="sc">~</span>  <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    sigma[d] <span class="sc">~</span>  <span class="fu">dunif</span>(<span class="dv">2</span><span class="sc">/</span><span class="dv">1000</span>, <span class="dv">2</span> <span class="sc">*</span> <span class="dv">1000</span>)  </span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    tau[d]   <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> sigma[d]<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>  delta_mu    <span class="ot">&lt;-</span> mu[<span class="dv">2</span>] <span class="sc">-</span> mu[<span class="dv">1</span>]</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>  delta_sigma <span class="ot">&lt;-</span> sigma[<span class="dv">2</span>] <span class="sc">-</span> sigma[<span class="dv">1</span>]</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>sleep_jags2 <span class="ot">&lt;-</span> </span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sleep_model2,</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"mu"</span>, <span class="st">"sigma"</span>, <span class="st">"delta_mu"</span>, <span class="st">"delta_sigma"</span>, <span class="st">"tau"</span>),</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">extra =</span> sleep<span class="sc">$</span>extra,</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">drug  =</span> sleep<span class="sc">$</span>drug,</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">Nobs  =</span> <span class="fu">nrow</span>(sleep),</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">Ndrugs =</span> <span class="dv">2</span></span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">DIC =</span> <span class="cn">FALSE</span></span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Resultados:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sleep_jags2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fit using jags
 3 chains, each with 2000 iterations (first 1000 discarded)
 n.sims = 3000 iterations saved
            mu.vect sd.vect   2.5%    25%   50%   75% 97.5%  Rhat n.eff
delta_mu      1.481   0.981 -0.553  0.837 1.513 2.122 3.369 1.001  3000
delta_sigma   0.268   0.891 -1.491 -0.241 0.231 0.765 2.124 1.006  3000
mu[1]         0.705   0.643 -0.587  0.308 0.710 1.117 1.963 1.001  3000
mu[2]         2.186   0.751  0.666  1.714 2.206 2.670 3.620 1.001  2600
sigma[1]      2.076   0.583  1.282  1.659 1.972 2.340 3.495 1.002  1900
sigma[2]      2.344   0.679  1.437  1.884 2.209 2.665 3.934 1.004  1000
tau[1]        0.283   0.141  0.082  0.183 0.257 0.364 0.609 1.002  1900
tau[2]        0.223   0.112  0.065  0.141 0.205 0.282 0.484 1.004  1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>sleep_mcmc2 <span class="ot">&lt;-</span> <span class="fu">as.mcmc</span>(sleep_jags2)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc2, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">regex_pars =</span> <span class="st">"mu"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc2, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">regex_pars =</span> <span class="st">"sigma"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Probabilidades posteriores:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prop</span>( <span class="sc">~</span>(delta_mu <span class="sc">&gt;</span> <span class="dv">0</span>), <span class="at">data =</span> <span class="fu">posterior</span>(sleep_jags2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>prop_TRUE 
0.9313333 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prop</span>( <span class="sc">~</span>(delta_sigma <span class="sc">&gt;</span> <span class="dv">0</span>), <span class="at">data =</span> <span class="fu">posterior</span>(sleep_jags2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>prop_TRUE 
0.6273333 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(sleep_jags2, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"delta"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] par   lo    hi    prob  chain
&lt;0 rows&gt; (or 0-length row.names)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(sleep_jags2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           par          lo        hi prob chain
1     delta_mu -0.31002813 3.5258596 0.95     1
2     delta_mu -0.58403537 3.4175623 0.95     2
3     delta_mu -0.46920553 3.3184710 0.95     3
4  delta_sigma -1.65884403 2.0419410 0.95     1
5  delta_sigma -1.49119963 2.2485091 0.95     2
6  delta_sigma -1.22031281 1.9648398 0.95     3
7        mu[1] -0.48699891 2.0134925 0.95     1
8        mu[1] -0.59781972 1.9963228 0.95     2
9        mu[1] -0.56654155 1.8347895 0.95     3
10       mu[2]  0.83941898 3.6885670 0.95     1
11       mu[2]  0.57439664 3.5707756 0.95     2
12       mu[2]  0.78891482 3.6594392 0.95     3
13    sigma[1]  1.16789343 3.4282741 0.95     1
14    sigma[1]  1.25074100 3.3422105 0.95     2
15    sigma[1]  1.17272244 3.0258651 0.95     3
16    sigma[2]  1.29912840 3.6246450 0.95     1
17    sigma[2]  1.30504880 3.8035711 0.95     2
18    sigma[2]  1.29832108 3.5145432 0.95     3
19      tau[1]  0.05591998 0.5561229 0.95     1
20      tau[1]  0.07962254 0.5898826 0.95     2
21      tau[1]  0.07993175 0.5637375 0.95     3
22      tau[2]  0.04658374 0.4517959 0.95     1
23      tau[2]  0.04414731 0.4505996 0.95     2
24      tau[2]  0.05781056 0.4468488 0.95     3</code></pre>
</div>
</div>
<p>Comparación con una prueba t de dos muestras:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(extra <span class="sc">~</span> drug, <span class="at">data =</span> sleep)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  extra by drug
t = -1.8608, df = 17.776, p-value = 0.07939
alternative hypothesis: true difference in means between group 1 and group 2 is not equal to 0
95 percent confidence interval:
 -3.3654832  0.2054832
sample estimates:
mean in group 1 mean in group 2 
           0.75            2.33 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prop</span>( <span class="sc">~</span>(delta_mu <span class="sc">&lt;</span> <span class="dv">0</span>), <span class="at">data =</span> <span class="fu">posterior</span>(sleep_jags2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> prop_TRUE 
0.06866667 </code></pre>
</div>
</div>
</section>
<section id="rope-región-de-equivalencia-práctica" class="level3" data-number="6.1.4">
<h3 data-number="6.1.4" class="anchored" data-anchor-id="rope-región-de-equivalencia-práctica"><span class="header-section-number">6.1.4</span> ROPE (Región de Equivalencia Práctica)</h3>
<p>Saber que dos cosas no son iguales no es de mucha utilidad práctica si la diferencia es pequeña. Una forma de cuantificar esto es especificar una <strong>región de equivalencia práctica</strong> (ROPE, por sus siglas en inglés). Podríamos decidir, por ejemplo, que no nos interesan las diferencias de menos de 10 minutos (1/6 horas). Nuestra ROPE (para la diferencia en medias) sería entonces el intervalo <span class="math inline">\((-1/6, 1/6)\)</span> y podríamos preguntarnos si hay evidencia de que la diferencia verdadera se encuentra fuera de ese intervalo. Esto podría verificarse viendo si un HDI (Intervalo de Alta Densidad) se encuentra completamente fuera de la ROPE.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_post</span>(<span class="fu">posterior</span>(sleep_jags2)<span class="sc">$</span>delta_mu, <span class="at">ROPE =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span><span class="sc">/</span><span class="dv">6</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">6</span>),</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">hdi_prob =</span> <span class="fl">0.9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$posterior
      ESS     mean   median     mode
var1 3000 1.481005 1.513207 1.734047

$hdi
  prob         lo      hi
1  0.9 -0.1688969 3.00691

$ROPE
          lo        hi  P(&lt; ROPE) P(in ROPE) P(&gt; ROPE)
1 -0.1666667 0.1666667 0.04866667 0.04333333     0.908</code></pre>
</div>
</div>
</section>
<section id="comparaciones-pareadas" class="level3" data-number="6.1.5">
<h3 data-number="6.1.5" class="anchored" data-anchor-id="comparaciones-pareadas"><span class="header-section-number">6.1.5</span> Comparaciones Pareadas</h3>
<p>Los datos en realidad contienen otra variable: <code>ID</code>. Resulta que las mismas diez personas fueron evaluadas con cada medicamento. Si estamos principalmente interesados en comparar los dos medicamentos, podríamos tomar la diferencia entre el sueño adicional con un medicamento y con el otro <strong>para cada persona</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>sleep_wide <span class="ot">&lt;-</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  datasets<span class="sc">::</span>sleep <span class="sc">%&gt;%</span> </span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">drug =</span> group) <span class="sc">%&gt;%</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">drug =</span> <span class="fu">paste0</span>(<span class="st">"drug"</span>, drug)) <span class="sc">%&gt;%</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">spread</span>(<span class="at">key =</span> drug, <span class="at">value =</span> extra) </span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>sleep_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID drug1 drug2
1   1   0.7   1.9
2   2  -1.6   0.8
3   3  -0.2   1.1
4   4  -1.2   0.1
5   5  -0.1  -0.1
6   6   3.4   4.4
7   7   3.7   5.5
8   8   0.8   1.6
9   9   0.0   4.6
10 10   2.0   3.4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>sleep_wide <span class="ot">&lt;-</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  sleep_wide <span class="sc">%&gt;%</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">delta =</span> drug2 <span class="sc">-</span> drug1)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>sleep_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID drug1 drug2 delta
1   1   0.7   1.9   1.2
2   2  -1.6   0.8   2.4
3   3  -0.2   1.1   1.3
4   4  -1.2   0.1   1.3
5   5  -0.1  -0.1   0.0
6   6   3.4   4.4   1.0
7   7   3.7   5.5   1.8
8   8   0.8   1.6   0.8
9   9   0.0   4.6   4.6
10 10   2.0   3.4   1.4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_boxplot</span>(<span class="sc">~</span> delta, <span class="at">data =</span> sleep_wide)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>sleep_model4 <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>Nsubj) {</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>    delta[i] <span class="sc">~</span> <span class="fu">dt</span>(mu, <span class="dv">1</span> <span class="sc">/</span> sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  mu         <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">2</span>)</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>  sigma      <span class="sc">~</span> <span class="fu">dunif</span>(<span class="dv">2</span><span class="sc">/</span><span class="dv">1000</span>, <span class="dv">2</span> <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">29</span>)</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  nu        <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>  tau       <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> sigma<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>sleep_jags4 <span class="ot">&lt;-</span> </span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sleep_model4,</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"mu"</span>, <span class="st">"sigma"</span>, <span class="st">"nu"</span>),</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">delta =</span> sleep_wide<span class="sc">$</span>delta,</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">Nsubj =</span> <span class="fu">nrow</span>(sleep_wide)</span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span>,</span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">DIC =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Compiling model graph
   Resolving undeclared variables
   Allocating nodes
Graph information:
   Observed stochastic nodes: 10
   Unobserved stochastic nodes: 3
   Total graph size: 25

Initializing model</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sleep_jags4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fit using jags
 3 chains, each with 5000 iterations (first 2500 discarded), n.thin = 2
 n.sims = 3750 iterations saved
      mu.vect sd.vect  2.5%   25%    50%    75%  97.5%  Rhat n.eff
mu      1.105   0.368 0.252 0.906  1.135  1.346  1.747 1.001  3800
nu     21.684  26.066 1.308 4.446 11.673 29.393 99.611 1.004   640
sigma   1.180   0.490 0.396 0.848  1.130  1.436  2.332 1.003   760</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>sleep_mcmc4 <span class="ot">&lt;-</span> <span class="fu">as.mcmc</span>(sleep_jags4)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc4, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">pars =</span> <span class="st">"mu"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(sleep_mcmc4, <span class="at">prob =</span> <span class="fl">0.95</span>, <span class="at">pars =</span> <span class="st">"nu"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_pairs</span>(sleep_mcmc4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-12-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prop</span>( <span class="sc">~</span>(mu <span class="sc">&gt;</span> <span class="dv">0</span>), <span class="at">data =</span> <span class="fu">posterior</span>(sleep_jags4))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>prop_TRUE 
0.9890667 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(sleep_jags4, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"mu"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  par        lo       hi prob chain
1  mu 0.3270903 1.799986 0.95     1
2  mu 0.3527186 1.782928 0.95     2
3  mu 0.3312891 1.753492 0.95     3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(sleep_jags4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    par        lo        hi prob chain
1    mu 0.3270903  1.799986 0.95     1
2    mu 0.3527186  1.782928 0.95     2
3    mu 0.3312891  1.753492 0.95     3
4    nu 1.0042849 79.192907 0.95     1
5    nu 1.0035093 77.405855 0.95     2
6    nu 1.0104257 68.979008 0.95     3
7 sigma 0.3021670  2.114909 0.95     1
8 sigma 0.2622193  2.010899 0.95     2
9 sigma 0.3099302  2.186333 0.95     3</code></pre>
</div>
</div>
</section>
</section>
<section id="regresión-lineal-simple" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="regresión-lineal-simple"><span class="header-section-number">6.2</span> Regresión Lineal Simple</h2>
<p>Carga de STAN:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstan)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: StanHeaders</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
rstan version 2.32.5 (Stan version 2.32.2)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For execution on a local, multicore CPU with excess RAM we recommend calling
options(mc.cores = parallel::detectCores()).
To avoid recompilation of unchanged Stan programs, we recommend calling
rstan_options(auto_write = TRUE)
For within-chain threading using `reduce_sum()` or `map_rect()` Stan functions,
change `threads_per_chain` option:
rstan_options(threads_per_chain = 1)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'rstan'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:R2jags':

    traceplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:coda':

    traceplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:tidyr':

    extract</code></pre>
</div>
</div>
<section id="ejemplo-datos-de-galton" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="ejemplo-datos-de-galton"><span class="header-section-number">6.2.1</span> Ejemplo: Datos de Galton</h3>
<p>Dado que estamos viendo regresión, usemos un conjunto de datos históricos que formó parte de los orígenes de la historia de la regresión: los datos de Galton sobre la altura. Galton recopiló datos sobre la altura de los adultos y sus padres.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mosaicData)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Galton)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  family father mother sex height nkids
1      1   78.5   67.0   M   73.2     4
2      1   78.5   67.0   F   69.2     4
3      1   78.5   67.0   F   69.0     4
4      1   78.5   67.0   F   69.0     4
5      2   75.5   66.5   M   73.5     4
6      2   75.5   66.5   M   72.5     4</code></pre>
</div>
</div>
<p>Para simplificar las cosas por el momento, consideremos solo a las mujeres y solo un hermano por familia.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">54321</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>GaltonW <span class="ot">&lt;-</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  mosaicData<span class="sc">::</span>Galton <span class="sc">%&gt;%</span> </span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(sex <span class="sc">==</span> <span class="st">"F"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(family) <span class="sc">%&gt;%</span></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Problema de estudio</strong>: Galton estaba interesado en cómo la altura de las personas se relaciona con la altura de sus padres. Combinó la altura de los padres en la “altura media de los padres”, que era el promedio de ambos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>GaltonW <span class="ot">&lt;-</span> </span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>  GaltonW <span class="sc">%&gt;%</span></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">midparent =</span> (father <span class="sc">+</span> mother) <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(height <span class="sc">~</span> midparent, <span class="at">data =</span> GaltonW, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="modelo-de-regresión" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="modelo-de-regresión"><span class="header-section-number">6.2.2</span> Modelo de regresión</h3>
<section id="verosimilitud" class="level4" data-number="6.2.2.1">
<h4 data-number="6.2.2.1" class="anchored" data-anchor-id="verosimilitud"><span class="header-section-number">6.2.2.1</span> Verosimilitud</h4>
<span class="math display">\[\begin{align*}
y_{i} &amp;\sim {\sf Norm}(\mu_i, \sigma) \\
\mu_i &amp;\sim \beta_0 + \beta_1 x_i
\end{align*}\]</span>
<p>Algunas variaciones:</p>
<ul>
<li>Sustituir la distribución normal por otra (t es común).</li>
<li>Permitir que las desviaciones estándar varíen con <span class="math inline">\(x\)</span> así como la media.</li>
<li>Utilizar una relación funcional diferente entre la variable explicativa y la respuesta (regresión no lineal).</li>
</ul>
<p>La primera variación a veces se llama <strong>regresión robusta</strong> porque es más robusta ante observaciones inusuales.</p>
<span class="math display">\[\begin{align*}
y_{i} &amp;\sim {\sf T}(\mu_i, \sigma, \nu) \\
\mu_i &amp;\sim \beta_0 + \beta_1 x_i
\end{align*}\]</span>
</section>
<section id="previas" class="level4" data-number="6.2.2.2">
<h4 data-number="6.2.2.2" class="anchored" data-anchor-id="previas"><span class="header-section-number">6.2.2.2</span> Previas</h4>
<p>Necesitamos previas para <span class="math inline">\(\beta_0\)</span>, <span class="math inline">\(\beta_1\)</span>, <span class="math inline">\(\sigma\)</span> y <span class="math inline">\(\nu\)</span>.</p>
<ul>
<li><p><span class="math inline">\(\nu\)</span>: Ya hemos visto que una <strong>Gamma trasladada</strong> con media alrededor de 30 funciona bien como una previa genérica, dando a los datos espacio para alejarnos de la normalidad si es necesario.</p></li>
<li><p><span class="math inline">\(\beta_1\)</span>: El MLE para <span class="math inline">\(\beta_1\)</span> es</p>
<p><span class="math display">\[ \hat\beta_1 = r \frac{SD_y}{SD_x}\]</span> por lo que tiene sentido tener una previa que cubra ampliamente el intervalo <span class="math inline">\((- \frac{SD_y}{SD_x}, \frac{SD_y}{SD_x})\)</span>.</p></li>
<li><p><span class="math inline">\(\beta_0\)</span>: El MLE para <span class="math inline">\(\beta_0\)</span> es</p>
<p><span class="math display">\[ \hat\beta_0 \; = \;  \overline{y} - \hat \beta_1 \overline{x}  \; = \; \overline{y} - r \frac{SD_y}{SD_x} \cdot \overline{x}\]</span> por lo que podemos elegir un prior que cubra ampliamente el intervalo <span class="math inline">\((\overline{y} - \frac{SD_y}{SD_x} \cdot \overline{x}, \overline{y} - \frac{SD_y}{SD_x} \cdot \overline{x})\)</span></p></li>
<li><p><span class="math inline">\(\sigma\)</span> mide la cantidad de variabilidad en las respuestas para un valor <em>fijo</em> de <span class="math inline">\(x\)</span>. Una previa débilmente informativa debería cubrir el rango de valores razonables de <span class="math inline">\(\sigma\)</span> con bastante margen.</p></li>
</ul>
<p>Implementación en JAGS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>galton_model <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(y)) {</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>    y[i]   <span class="sc">~</span> <span class="fu">dt</span>(mu[i], <span class="dv">1</span><span class="sc">/</span>sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>    mu[i] <span class="ot">&lt;-</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> x[i]</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  sigma <span class="sc">~</span> <span class="fu">dunif</span>(<span class="dv">6</span><span class="sc">/</span><span class="dv">100</span>, <span class="dv">6</span> <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">29</span>)</span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>  nu <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a>  beta0 <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span><span class="sc">^</span><span class="dv">2</span>)   <span class="co"># 100 is order of magnitude of data</span></span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a>  beta1 <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span><span class="sc">^</span><span class="dv">2</span>)     <span class="co"># expect roughly 1-1 slope</span></span>
<span id="cb96-11"><a href="#cb96-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>galton_jags <span class="ot">&lt;-</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> galton_model,</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(<span class="at">y =</span> GaltonW<span class="sc">$</span>height, <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent),</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"beta0"</span>, <span class="st">"beta1"</span>, <span class="st">"sigma"</span>, <span class="st">"nu"</span>),</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span>,</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.burnin =</span> <span class="dv">2000</span>,</span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">4</span>,</span>
<span id="cb97-9"><a href="#cb97-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.thin =</span> <span class="dv">1</span></span>
<span id="cb97-10"><a href="#cb97-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Compiling model graph
   Resolving undeclared variables
   Allocating nodes
Graph information:
   Observed stochastic nodes: 168
   Unobserved stochastic nodes: 4
   Total graph size: 439

Initializing model</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(galton_jags)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fit using jags
 4 chains, each with 5000 iterations (first 2000 discarded)
 n.sims = 12000 iterations saved
         mu.vect sd.vect    2.5%     25%     50%     75%   97.5%  Rhat n.eff
beta0     20.485  10.628   4.679  11.691  19.836  28.017  43.115 2.443     5
beta1      0.654   0.159   0.314   0.541   0.663   0.785   0.890 2.825     5
nu        37.706  29.592   7.104  16.803  28.849  49.512 119.715 1.001  4000
sigma      1.956   0.133   1.703   1.865   1.955   2.044   2.222 1.023   110
deviance 715.427   5.446 709.847 711.779 713.932 717.038 732.172 1.657     9

DIC info (using the rule, pD = var(deviance)/2)
pD = 9.1 and DIC = 724.5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc</span>(galton_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Claramente la convergencia no es la más adecuada.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior</span>(galton_jags) <span class="sc">%&gt;%</span> </span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(beta0 <span class="sc">~</span> beta1, <span class="at">color =</span> <span class="sc">~</span> chain, <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">size =</span> <span class="fl">0.4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_density2d</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch17-galton-problems-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior</span>(galton_jags) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(iter <span class="sc">&lt;=</span> <span class="dv">250</span>, chain <span class="sc">==</span> <span class="st">"chain:1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_step</span>(beta0 <span class="sc">~</span> beta1, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="sc">~</span>iter) <span class="sc">%&gt;%</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_density2d</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">%&gt;%</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_refine</span>(<span class="fu">scale_color_viridis_c</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_facet_wrap</span>(<span class="sc">~</span>chain) <span class="co">#, scales = "free")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following aesthetics were dropped during statistical transformation:
colour.
ℹ This can happen when ggplot fails to infer the correct grouping structure in
  the data.
ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
  variable into a factor?</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch17-galton-problems-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>La correlación de los parámetros en la distribución posterior produce una cresta larga, estrecha y diagonal que el muestreador de Gibbs muestrea muy lentamente porque sigue chocando con el borde de la región.</p>
<p>Entonces, ¿cómo solucionamos esto? Se supone que este es el modelo lineal <em>simple</em> después de todo. Hay dos formas en las que podríamos esperar solucionar nuestro problema.</p>
<ol type="1">
<li><p><strong>Reparametrizar el modelo</strong> para que la correlación entre los parámetros (en la distribución posterior) se reduzca o elimine.</p></li>
<li><p><strong>Usar un algoritmo diferente</strong> para el muestreo posterior.</p></li>
</ol>
<p>Reparametrización 1: <strong>centrado</strong></p>
<p>Podemos expresar este modelo como</p>
<span class="math display">\[\begin{align*}
y_{i} &amp;\sim {\sf T}(\mu_i, \sigma, \nu) \\
\mu_i &amp;= \alpha_0 + \alpha_1 (x_i - \overline{x})
\end{align*}\]</span>
<p>Dado que</p>
<span class="math display">\[\begin{align*}
\alpha_0 + \alpha_1 (x_i - \overline{x})
&amp;= (\alpha_0 - \alpha_1 \overline{x}) + \alpha_1 x_i
\end{align*}\]</span>
<p>Vemos que <span class="math inline">\(\beta_0 = \alpha_0 - \alpha_1 \overline{x}\)</span> y <span class="math inline">\(\beta_1 = \alpha_1\)</span>. Por lo tanto, podemos recuperar fácilmente los parámetros originales si lo deseamos. (Y si estamos principalmente interesados en <span class="math inline">\(\beta_1\)</span>, no se requiere ninguna traducción).</p>
<p>Esta reparametrización mantiene la escala natural de los datos, y tanto <span class="math inline">\(\alpha_0\)</span> como <span class="math inline">\(\alpha_1\)</span> son fácilmente interpretados: <span class="math inline">\(\alpha_0\)</span> es la respuesta media cuando el predictor es el promedio de los valores del predictor <em>en los datos</em>.</p>
<p>Modelación en JAGS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>galtonC_model <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(y)) {</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>    y[i]   <span class="sc">~</span> <span class="fu">dt</span>(mu[i], <span class="dv">1</span><span class="sc">/</span>sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>    mu[i] <span class="ot">&lt;-</span> alpha0 <span class="sc">+</span> alpha1 <span class="sc">*</span> (x[i] <span class="sc">-</span> <span class="fu">mean</span>(x))</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>  sigma <span class="sc">~</span> <span class="fu">dunif</span>(<span class="dv">6</span><span class="sc">/</span><span class="dv">100</span>, <span class="dv">6</span> <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">29</span>)</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>  nu <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>  alpha0 <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span><span class="sc">^</span><span class="dv">2</span>)   <span class="co"># 100 is order of magnitude of data</span></span>
<span id="cb105-10"><a href="#cb105-10" aria-hidden="true" tabindex="-1"></a>  alpha1 <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span><span class="sc">^</span><span class="dv">2</span>)     <span class="co"># expect roughly 1-1 slope</span></span>
<span id="cb105-11"><a href="#cb105-11" aria-hidden="true" tabindex="-1"></a>  beta0 <span class="ot">=</span> alpha0 <span class="sc">-</span> alpha1 <span class="sc">*</span> <span class="fu">mean</span>(x)</span>
<span id="cb105-12"><a href="#cb105-12" aria-hidden="true" tabindex="-1"></a>  beta1 <span class="ot">=</span> alpha1               <span class="co"># not necessary, but gives us both names</span></span>
<span id="cb105-13"><a href="#cb105-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb105-14"><a href="#cb105-14" aria-hidden="true" tabindex="-1"></a>galtonC_jags <span class="ot">&lt;-</span></span>
<span id="cb105-15"><a href="#cb105-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb105-16"><a href="#cb105-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> galtonC_model,</span>
<span id="cb105-17"><a href="#cb105-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(<span class="at">y =</span> GaltonW<span class="sc">$</span>height, <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent),</span>
<span id="cb105-18"><a href="#cb105-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"beta0"</span>, <span class="st">"beta1"</span>, <span class="st">"alpha0"</span>, <span class="st">"alpha1"</span>, <span class="st">"sigma"</span>, <span class="st">"nu"</span>),</span>
<span id="cb105-19"><a href="#cb105-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span>,</span>
<span id="cb105-20"><a href="#cb105-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.burnin =</span> <span class="dv">2000</span>,</span>
<span id="cb105-21"><a href="#cb105-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">4</span>,</span>
<span id="cb105-22"><a href="#cb105-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.thin =</span> <span class="dv">1</span></span>
<span id="cb105-23"><a href="#cb105-23" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Compiling model graph
   Resolving undeclared variables
   Allocating nodes
Graph information:
   Observed stochastic nodes: 168
   Unobserved stochastic nodes: 4
   Total graph size: 484

Initializing model</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(galtonC_jags)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>fit using jags
 4 chains, each with 5000 iterations (first 2000 discarded)
 n.sims = 12000 iterations saved
         mu.vect sd.vect    2.5%     25%     50%     75%   97.5%  Rhat n.eff
alpha0    64.131   0.157  63.824  64.026  64.130  64.237  64.443 1.001  5800
alpha1     0.706   0.085   0.539   0.649   0.706   0.763   0.873 1.001 12000
beta0     16.964   5.662   5.816  13.161  16.985  20.773  28.140 1.001 12000
beta1      0.706   0.085   0.539   0.649   0.706   0.763   0.873 1.001 12000
nu        37.384  29.036   7.126  17.277  28.864  48.228 117.040 1.003  1200
sigma      1.940   0.128   1.686   1.856   1.940   2.026   2.190 1.002  3200
deviance 712.626   2.596 709.617 710.731 711.961 713.813 719.426 1.002  2400

DIC info (using the rule, pD = var(deviance)/2)
pD = 3.4 and DIC = 716.0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc</span>(galtonC_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(beta1 <span class="sc">~</span> beta0, <span class="at">data =</span> <span class="fu">posterior</span>(galtonC_jags), <span class="at">alpha =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(alpha1 <span class="sc">~</span> alpha0, <span class="at">data =</span> <span class="fu">posterior</span>(galtonC_jags), <span class="at">alpha =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="análisis-posterior" class="level3" data-number="6.2.3">
<h3 data-number="6.2.3" class="anchored" data-anchor-id="análisis-posterior"><span class="header-section-number">6.2.3</span> Análisis posterior</h3>
<section id="estimar-parámetros" class="level4" data-number="6.2.3.1">
<h4 data-number="6.2.3.1" class="anchored" data-anchor-id="estimar-parámetros"><span class="header-section-number">6.2.3.1</span> Estimar parámetros</h4>
<p>Si estamos principalmente interesados en un parámetro de regresión (generalmente el parámetro de pendiente es mucho más interesante que el parámetro de intercepción), podemos usar un HDI para expresar nuestra estimación.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(galtonC_jags), <span class="at">pars =</span> <span class="st">"beta1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    par        lo        hi      mode prob
1 beta1 0.5435323 0.8770839 0.7127963 0.95</code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(<span class="fu">as.mcmc</span>(galtonC_jags), <span class="at">pars =</span> <span class="st">"beta1"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch17-galtonC-estimate-params-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Galton notó lo que vemos aquí: que la pendiente es menor que 1. Esto significa que los hijos de padres más altos que el promedio tienden a ser más bajos que sus padres y los hijos de padres por debajo del promedio tienden a ser más altos que sus padres. Se refirió a esto en su artículo como [“regresión hacia la mediocridad”]. Resulta que esto no fue una característica especial de la herencia de la altura, sino una característica general de los modelos lineales.</p>
</section>
<section id="hacer-predicciones" class="level4" data-number="6.2.3.2">
<h4 data-number="6.2.3.2" class="anchored" data-anchor-id="hacer-predicciones"><span class="header-section-number">6.2.3.2</span> Hacer predicciones</h4>
<p>Supongamos que conocemos las alturas de un padre y una madre, a partir de las cuales calculamos la altura del padre y la madre (<span class="math inline">\(x\)</span>). ¿Qué tan alto predeciríamos que serán sus hijas cuando sean adultas? Cada muestra posterior proporciona una respuesta describiendo una distribución t con <code>nu</code> grados de libertad, media <span class="math inline">\(\beta_0 + \beta_1 x\)</span> y desviación estándar <span class="math inline">\(\sigma\)</span>.</p>
<p>La distribución posterior de la altura promedio de las hijas nacidas de padres con una altura de padre y madre de <span class="math inline">\(x = 70\)</span> se muestra a continuación, junto con un HDI.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior</span>(galtonC_jags) <span class="sc">%&gt;%</span> </span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mean_daughter =</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> <span class="dv">70</span>) <span class="sc">%&gt;%</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_dens</span>(<span class="sc">~</span>mean_daughter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>Galton_hdi <span class="ot">&lt;-</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior</span>(galtonC_jags) <span class="sc">%&gt;%</span> </span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mean_daughter =</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> <span class="dv">70</span>) <span class="sc">%&gt;%</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hdi</span>(<span class="at">pars =</span> <span class="st">"mean_daughter"</span>)</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>Galton_hdi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            par       lo       hi     mode prob
1 mean_daughter 65.77185 67.01363 66.34311 0.95</code></pre>
</div>
</div>
<p>Entonces, en promedio, predeciríamos que las hijas tienen aproximadamente 66 o 67 pulgadas de altura.</p>
<p>Podemos visualizar esto dibujando una línea para cada muestra posterior. El HDI debería abarcar el 95% intermedio de estos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_abline</span>(<span class="at">intercept =</span> <span class="sc">~</span>beta0, <span class="at">slope =</span> <span class="sc">~</span>beta1, <span class="at">alpha =</span> <span class="fl">0.01</span>,</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> <span class="st">"steelblue"</span>, </span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">data =</span> <span class="fu">posterior</span>(galtonC_jags) <span class="sc">%&gt;%</span> <span class="fu">sample_n</span>(<span class="dv">2000</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(height <span class="sc">~</span> midparent, <span class="at">data =</span> GaltonW, </span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">inherit =</span> <span class="cn">FALSE</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_errorbar</span>(lo <span class="sc">+</span> hi <span class="sc">~</span> <span class="dv">70</span>, <span class="at">data =</span> Galton_hdi, <span class="at">color =</span> <span class="st">"skyblue"</span>, </span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">size =</span> <span class="fl">1.2</span>, <span class="at">inherit =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Pero esta puede no ser el tipo de predicción que queremos. Observa que las alturas de la mayoría de las hijas no están dentro de la banda azul en la imagen. Esa banda habla sobre la <em>media</em>, pero no tiene en cuenta cuánto varían los individuos alrededor de esa media.</p>
<p>Aquí generamos alturas agregando ruido a la estimación dada por los valores de <span class="math inline">\(\beta_0\)</span> y <span class="math inline">\(\beta_1\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior</span>(galtonC_jags) <span class="sc">%&gt;%</span>  </span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">new_ht =</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> <span class="dv">70</span> <span class="sc">+</span> <span class="fu">rt</span>(<span class="dv">1200</span>, <span class="at">df =</span> nu) <span class="sc">*</span> sigma) <span class="sc">%&gt;%</span></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(new_ht <span class="sc">~</span> <span class="dv">70</span>, <span class="at">alpha =</span> <span class="fl">0.01</span>, <span class="at">size =</span> <span class="fl">0.7</span>, <span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(height <span class="sc">~</span> midparent, <span class="at">data =</span> GaltonW, </span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">inherit =</span> <span class="cn">FALSE</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>Galton_hdi2 <span class="ot">&lt;-</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior</span>(galtonC_jags) <span class="sc">%&gt;%</span> </span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">new_ht =</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> <span class="dv">70</span> <span class="sc">+</span> <span class="fu">rt</span>(<span class="dv">1200</span>, <span class="at">df =</span> nu) <span class="sc">*</span> sigma) <span class="sc">%&gt;%</span></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hdi</span>(<span class="at">regex_pars =</span> <span class="st">"new"</span>) </span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>Galton_hdi2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     par       lo       hi     mode prob
1 new_ht 62.13155 70.25104 65.91627 0.95</code></pre>
</div>
</div>
</section>
<section id="verificación-predictiva-posterior-con-bayesplot" class="level4" data-number="6.2.3.3">
<h4 data-number="6.2.3.3" class="anchored" data-anchor-id="verificación-predictiva-posterior-con-bayesplot"><span class="header-section-number">6.2.3.3</span> Verificación predictiva posterior con bayesplot</h4>
<p>El paquete bayesplot proporciona varios gráficos de verificación predictiva posterior (ppc). Estas funciones requieren dos entradas importantes:</p>
<ul>
<li><code>y</code>: un vector de valores de respuesta, generalmente los valores del conjunto de datos original.</li>
<li><code>yrep</code>: una matriz de valores <code>y</code> simulados. Cada fila corresponde a una muestra posterior. Hay una columna para cada valor de <code>y</code>.</li>
</ul>
<p>Entonces, las filas de <code>yrep</code> se pueden comparar con <code>y</code> para ver si el modelo se comporta bien.</p>
<p>Nota: Podemos calcular nuestros valores simulados <span class="math inline">\(y\)</span> utilizando valores de predictores que sean similares a los de nuestros datos o utilizando otros valores de predictores que elijamos. La segunda opción nos permite considerar situaciones contrafactuales. Para distinguir estos, algunas personas usan <span class="math inline">\(y_rep\)</span> para lo primero y <span class="math inline">\(\tilde{y}\)</span> para lo segundo.</p>
<p>Ahora todo el trabajo está en crear la matriz <code>yrep</code>. Para simplificar eso, usaremos <code>CalvinBayes::posterior_calc()</code>. Lo haremos de dos maneras, una vez para los valores promedio de altura y otra vez para los valores individuales de altura (teniendo en cuenta la variabilidad de persona a persona según lo cuantificado por <span class="math inline">\(\nu\)</span> y <span class="math inline">\(\sigma\)</span>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>y_avg <span class="ot">&lt;-</span> </span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior_calc</span>(</span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>    galtonC_jags, </span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>    height <span class="sc">~</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> midparent, </span>
<span id="cb123-5"><a href="#cb123-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> GaltonW)</span>
<span id="cb123-6"><a href="#cb123-6" aria-hidden="true" tabindex="-1"></a>y_ind <span class="ot">&lt;-</span> </span>
<span id="cb123-7"><a href="#cb123-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior_calc</span>(</span>
<span id="cb123-8"><a href="#cb123-8" aria-hidden="true" tabindex="-1"></a>    galtonC_jags, </span>
<span id="cb123-9"><a href="#cb123-9" aria-hidden="true" tabindex="-1"></a>    height <span class="sc">~</span> </span>
<span id="cb123-10"><a href="#cb123-10" aria-hidden="true" tabindex="-1"></a>      beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> midparent <span class="sc">+</span> <span class="fu">rt</span>(<span class="fu">nrow</span>(GaltonW), <span class="at">df =</span> nu) <span class="sc">*</span> sigma, </span>
<span id="cb123-11"><a href="#cb123-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> GaltonW)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Los diferentes gráficos de verificación predictiva posterior comienzan con <code>ppc_</code>. Aquí tienes un ejemplo:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ppc_intervals</span>(GaltonW<span class="sc">$</span>height, y_avg, <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ppc_intervals</span>(GaltonW<span class="sc">$</span>height, y_ind, <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-28-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Podemos extraer los datos utilizados para crear el gráfico y hacer nuestro propio gráfico como queramos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> </span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ppc_ribbon_data</span>(GaltonW<span class="sc">$</span>height, y_ind, <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent)</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(plot_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 168
Columns: 10
$ y_id        &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17,…
$ y_obs       &lt;dbl&gt; 69.0, 65.5, 62.5, 61.0, 64.0, 67.5, 65.5, 65.5, 64.0, 65.2…
$ x           &lt;dbl&gt; 72.75, 69.75, 67.50, 67.75, 68.00, 67.75, 67.75, 67.50, 67…
$ outer_width &lt;dbl&gt; 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9…
$ inner_width &lt;dbl&gt; 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5…
$ ll          &lt;dbl&gt; 64.87028, 62.84453, 61.28324, 61.40978, 61.65842, 61.46755…
$ l           &lt;dbl&gt; 66.95325, 64.86351, 63.28902, 63.46770, 63.64949, 63.46092…
$ m           &lt;dbl&gt; 68.37181, 66.19737, 64.62360, 64.81006, 64.99056, 64.81988…
$ h           &lt;dbl&gt; 69.72746, 67.53690, 65.96373, 66.16710, 66.34426, 66.18309…
$ hh          &lt;dbl&gt; 71.80631, 69.54626, 67.94110, 68.18508, 68.40763, 68.18462…</code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="sc">%&gt;%</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_ribbon</span>(ll <span class="sc">+</span> hh <span class="sc">~</span> x, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_ribbon</span>(l <span class="sc">+</span> h <span class="sc">~</span> x, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_line</span>(m <span class="sc">~</span> x, <span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(y_obs <span class="sc">~</span> x, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="sc">%&gt;%</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_smooth</span>(ll <span class="sc">~</span> x, <span class="at">color =</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_smooth</span>(hh <span class="sc">~</span> x, <span class="at">color=</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_smooth</span>(m <span class="sc">~</span> x, <span class="at">color=</span> <span class="st">"steelblue"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(y_obs <span class="sc">~</span> x, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using method = 'loess'
`geom_smooth()` using method = 'loess'
`geom_smooth()` using method = 'loess'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-29-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="ajuste-de-modelos-con-stan" class="level3" data-number="6.2.4">
<h3 data-number="6.2.4" class="anchored" data-anchor-id="ajuste-de-modelos-con-stan"><span class="header-section-number">6.2.4</span> Ajuste de modelos con Stan</h3>
<p>Centrar (o estandarizar) es suficiente para hacer que JAGS sea lo suficientemente eficiente como para usarlo. Pero también podemos usar Stan, y dado que Stan no se ve afectado por la correlación en el posterior de la manera en que lo hace JAGS, Stan funciona bien incluso sin reparametrizar el modelo.</p>
<p>El código correspondiente del modelo anterior está en el archivo <em>galton.stan</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>modelo_galton <span class="ot">&lt;-</span> <span class="fu">stan_model</span>(<span class="at">file =</span> <span class="st">'galton.stan'</span>,<span class="at">verbose =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
TRANSLATING MODEL '' FROM Stan CODE TO C++ CODE NOW.
OS: x86_64, linux-gnu; rstan: 2.32.5; Rcpp: 1.0.12; inline: 0.3.19 
 &gt;&gt; setting environment variables: 
PKG_LIBS =  '/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/rstan/lib//libStanServices.a' -L'/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/lib/' -lStanHeaders -L'/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppParallel/lib/' -ltbb 
PKG_CPPFLAGS =   -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/Rcpp/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppEigen/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppEigen/include/unsupported"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/BH/include" -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/src/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppParallel/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/rstan/include" -DEIGEN_NO_DEBUG  -DBOOST_DISABLE_ASSERTS  -DBOOST_PENDING_INTEGER_LOG2_HPP  -DSTAN_THREADS  -DUSE_STANC3 -DSTRICT_R_HEADERS  -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION  -D_HAS_AUTO_PTR_ETC=0  -include '/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/stan/math/prim/fun/Eigen.hpp'  -D_REENTRANT -DRCPP_PARALLEL_USE_TBB=1 
 &gt;&gt; Program source :

   1 : 
   2 : // includes from the plugin
   3 : // [[Rcpp::plugins(cpp14)]]
   4 : 
   5 : 
   6 : // user includes
   7 : #include &lt;Rcpp.h&gt;
   8 : using namespace Rcpp;
   9 : #ifndef MODELS_HPP
  10 : #define MODELS_HPP
  11 : #define STAN__SERVICES__COMMAND_HPP
  12 : #include &lt;rstan/rstaninc.hpp&gt;
  13 : #ifndef USE_STANC3
  14 : #define USE_STANC3
  15 : #endif
  16 : // Code generated by stanc v2.32.2
  17 : #include &lt;stan/model/model_header.hpp&gt;
  18 : namespace model5ad42f2ccaea__namespace {
  19 : using stan::model::model_base_crtp;
  20 : using namespace stan::math;
  21 : stan::math::profile_map profiles__;
  22 : static constexpr std::array&lt;const char*, 17&gt; locations_array__ =
  23 :   {" (found before start of program)",
  24 :   " (in 'anon_model', line 20, column 2 to column 13)",
  25 :   " (in 'anon_model', line 21, column 2 to column 13)",
  26 :   " (in 'anon_model', line 22, column 2 to column 22)",
  27 :   " (in 'anon_model', line 23, column 2 to column 27)",
  28 :   " (in 'anon_model', line 26, column 2 to column 19)",
  29 :   " (in 'anon_model', line 27, column 2 to column 22)",
  30 :   " (in 'anon_model', line 33, column 2 to column 46)",
  31 :   " (in 'anon_model', line 34, column 2 to column 25)",
  32 :   " (in 'anon_model', line 35, column 2 to column 23)",
  33 :   " (in 'anon_model', line 36, column 2 to column 44)",
  34 :   " (in 'anon_model', line 37, column 2 to column 35)",
  35 :   " (in 'anon_model', line 13, column 2 to column 17)",
  36 :   " (in 'anon_model', line 14, column 9 to column 10)",
  37 :   " (in 'anon_model', line 14, column 2 to column 14)",
  38 :   " (in 'anon_model', line 15, column 9 to column 10)",
  39 :   " (in 'anon_model', line 15, column 2 to column 14)"};
  40 : class model5ad42f2ccaea_ final : public model_base_crtp&lt;model5ad42f2ccaea_&gt; {
  41 : private:
  42 :   int N;
  43 :   Eigen::Matrix&lt;double,-1,1&gt; y_data__;
  44 :   Eigen::Matrix&lt;double,-1,1&gt; x_data__;
  45 :   Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt; y{nullptr, 0};
  46 :   Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt; x{nullptr, 0};
  47 : public:
  48 :   ~model5ad42f2ccaea_() {}
  49 :   model5ad42f2ccaea_(stan::io::var_context&amp; context__, unsigned int
  50 :                      random_seed__ = 0, std::ostream* pstream__ = nullptr)
  51 :       : model_base_crtp(0) {
  52 :     int current_statement__ = 0;
  53 :     using local_scalar_t__ = double;
  54 :     boost::ecuyer1988 base_rng__ =
  55 :       stan::services::util::create_rng(random_seed__, 0);
  56 :     // suppress unused var warning
  57 :     (void) base_rng__;
  58 :     static constexpr const char* function__ =
  59 :       "model5ad42f2ccaea__namespace::model5ad42f2ccaea_";
  60 :     // suppress unused var warning
  61 :     (void) function__;
  62 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
  63 :     // suppress unused var warning
  64 :     (void) DUMMY_VAR__;
  65 :     try {
  66 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
  67 :       pos__ = 1;
  68 :       current_statement__ = 12;
  69 :       context__.validate_dims("data initialization", "N", "int",
  70 :         std::vector&lt;size_t&gt;{});
  71 :       N = std::numeric_limits&lt;int&gt;::min();
  72 :       current_statement__ = 12;
  73 :       N = context__.vals_i("N")[(1 - 1)];
  74 :       current_statement__ = 12;
  75 :       stan::math::check_greater_or_equal(function__, "N", N, 0);
  76 :       current_statement__ = 13;
  77 :       stan::math::validate_non_negative_index("y", "N", N);
  78 :       current_statement__ = 14;
  79 :       context__.validate_dims("data initialization", "y", "double",
  80 :         std::vector&lt;size_t&gt;{static_cast&lt;size_t&gt;(N)});
  81 :       y_data__ = Eigen::Matrix&lt;double,-1,1&gt;::Constant(N,
  82 :                    std::numeric_limits&lt;double&gt;::quiet_NaN());
  83 :       new (&amp;y) Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(y_data__.data(), N);
  84 :       {
  85 :         std::vector&lt;local_scalar_t__&gt; y_flat__;
  86 :         current_statement__ = 14;
  87 :         y_flat__ = context__.vals_r("y");
  88 :         current_statement__ = 14;
  89 :         pos__ = 1;
  90 :         current_statement__ = 14;
  91 :         for (int sym1__ = 1; sym1__ &lt;= N; ++sym1__) {
  92 :           current_statement__ = 14;
  93 :           stan::model::assign(y, y_flat__[(pos__ - 1)],
  94 :             "assigning variable y", stan::model::index_uni(sym1__));
  95 :           current_statement__ = 14;
  96 :           pos__ = (pos__ + 1);
  97 :         }
  98 :       }
  99 :       current_statement__ = 15;
 100 :       stan::math::validate_non_negative_index("x", "N", N);
 101 :       current_statement__ = 16;
 102 :       context__.validate_dims("data initialization", "x", "double",
 103 :         std::vector&lt;size_t&gt;{static_cast&lt;size_t&gt;(N)});
 104 :       x_data__ = Eigen::Matrix&lt;double,-1,1&gt;::Constant(N,
 105 :                    std::numeric_limits&lt;double&gt;::quiet_NaN());
 106 :       new (&amp;x) Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(x_data__.data(), N);
 107 :       {
 108 :         std::vector&lt;local_scalar_t__&gt; x_flat__;
 109 :         current_statement__ = 16;
 110 :         x_flat__ = context__.vals_r("x");
 111 :         current_statement__ = 16;
 112 :         pos__ = 1;
 113 :         current_statement__ = 16;
 114 :         for (int sym1__ = 1; sym1__ &lt;= N; ++sym1__) {
 115 :           current_statement__ = 16;
 116 :           stan::model::assign(x, x_flat__[(pos__ - 1)],
 117 :             "assigning variable x", stan::model::index_uni(sym1__));
 118 :           current_statement__ = 16;
 119 :           pos__ = (pos__ + 1);
 120 :         }
 121 :       }
 122 :     } catch (const std::exception&amp; e) {
 123 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 124 :     }
 125 :     num_params_r__ = 1 + 1 + 1 + 1;
 126 :   }
 127 :   inline std::string model_name() const final {
 128 :     return "model5ad42f2ccaea_";
 129 :   }
 130 :   inline std::vector&lt;std::string&gt; model_compile_info() const noexcept {
 131 :     return std::vector&lt;std::string&gt;{"stanc_version = stanc3 v2.32.2",
 132 :              "stancflags = --"};
 133 :   }
 134 :   template &lt;bool propto__, bool jacobian__, typename VecR, typename VecI,
 135 :             stan::require_vector_like_t&lt;VecR&gt;* = nullptr,
 136 :             stan::require_vector_like_vt&lt;std::is_integral, VecI&gt;* = nullptr&gt;
 137 :   inline stan::scalar_type_t&lt;VecR&gt;
 138 :   log_prob_impl(VecR&amp; params_r__, VecI&amp; params_i__, std::ostream*
 139 :                 pstream__ = nullptr) const {
 140 :     using T__ = stan::scalar_type_t&lt;VecR&gt;;
 141 :     using local_scalar_t__ = T__;
 142 :     T__ lp__(0.0);
 143 :     stan::math::accumulator&lt;T__&gt; lp_accum__;
 144 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 145 :     int current_statement__ = 0;
 146 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 147 :     // suppress unused var warning
 148 :     (void) DUMMY_VAR__;
 149 :     static constexpr const char* function__ =
 150 :       "model5ad42f2ccaea__namespace::log_prob";
 151 :     // suppress unused var warning
 152 :     (void) function__;
 153 :     try {
 154 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 155 :       current_statement__ = 1;
 156 :       beta0 = in__.template read&lt;local_scalar_t__&gt;();
 157 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 158 :       current_statement__ = 2;
 159 :       beta1 = in__.template read&lt;local_scalar_t__&gt;();
 160 :       local_scalar_t__ sigma = DUMMY_VAR__;
 161 :       current_statement__ = 3;
 162 :       sigma = in__.template read_constrain_lb&lt;local_scalar_t__,
 163 :                 jacobian__&gt;(0, lp__);
 164 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 165 :       current_statement__ = 4;
 166 :       nuMinusOne = in__.template read_constrain_lb&lt;local_scalar_t__,
 167 :                      jacobian__&gt;(0, lp__);
 168 :       local_scalar_t__ nu = DUMMY_VAR__;
 169 :       current_statement__ = 6;
 170 :       nu = (nuMinusOne + 1);
 171 :       current_statement__ = 5;
 172 :       stan::math::check_greater_or_equal(function__, "nu", nu, 0);
 173 :       {
 174 :         current_statement__ = 7;
 175 :         lp_accum__.add(stan::math::student_t_lpdf&lt;propto__&gt;(y, nu,
 176 :                          stan::math::add(beta0,
 177 :                            stan::math::multiply(beta1, x)), sigma));
 178 :         current_statement__ = 8;
 179 :         lp_accum__.add(stan::math::normal_lpdf&lt;propto__&gt;(beta0, 0, 100));
 180 :         current_statement__ = 9;
 181 :         lp_accum__.add(stan::math::normal_lpdf&lt;propto__&gt;(beta1, 0, 4));
 182 :         current_statement__ = 10;
 183 :         lp_accum__.add(stan::math::uniform_lpdf&lt;propto__&gt;(sigma, (6.0 /
 184 :                          100.0), (6.0 * 100.0)));
 185 :         current_statement__ = 11;
 186 :         lp_accum__.add(stan::math::exponential_lpdf&lt;propto__&gt;(nuMinusOne, (1
 187 :                          / 29.0)));
 188 :       }
 189 :     } catch (const std::exception&amp; e) {
 190 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 191 :     }
 192 :     lp_accum__.add(lp__);
 193 :     return lp_accum__.sum();
 194 :   }
 195 :   template &lt;typename RNG, typename VecR, typename VecI, typename VecVar,
 196 :             stan::require_vector_like_vt&lt;std::is_floating_point,
 197 :             VecR&gt;* = nullptr, stan::require_vector_like_vt&lt;std::is_integral,
 198 :             VecI&gt;* = nullptr, stan::require_vector_vt&lt;std::is_floating_point,
 199 :             VecVar&gt;* = nullptr&gt;
 200 :   inline void
 201 :   write_array_impl(RNG&amp; base_rng__, VecR&amp; params_r__, VecI&amp; params_i__,
 202 :                    VecVar&amp; vars__, const bool
 203 :                    emit_transformed_parameters__ = true, const bool
 204 :                    emit_generated_quantities__ = true, std::ostream*
 205 :                    pstream__ = nullptr) const {
 206 :     using local_scalar_t__ = double;
 207 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 208 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 209 :     static constexpr bool propto__ = true;
 210 :     // suppress unused var warning
 211 :     (void) propto__;
 212 :     double lp__ = 0.0;
 213 :     // suppress unused var warning
 214 :     (void) lp__;
 215 :     int current_statement__ = 0;
 216 :     stan::math::accumulator&lt;double&gt; lp_accum__;
 217 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 218 :     // suppress unused var warning
 219 :     (void) DUMMY_VAR__;
 220 :     constexpr bool jacobian__ = false;
 221 :     static constexpr const char* function__ =
 222 :       "model5ad42f2ccaea__namespace::write_array";
 223 :     // suppress unused var warning
 224 :     (void) function__;
 225 :     try {
 226 :       double beta0 = std::numeric_limits&lt;double&gt;::quiet_NaN();
 227 :       current_statement__ = 1;
 228 :       beta0 = in__.template read&lt;local_scalar_t__&gt;();
 229 :       double beta1 = std::numeric_limits&lt;double&gt;::quiet_NaN();
 230 :       current_statement__ = 2;
 231 :       beta1 = in__.template read&lt;local_scalar_t__&gt;();
 232 :       double sigma = std::numeric_limits&lt;double&gt;::quiet_NaN();
 233 :       current_statement__ = 3;
 234 :       sigma = in__.template read_constrain_lb&lt;local_scalar_t__,
 235 :                 jacobian__&gt;(0, lp__);
 236 :       double nuMinusOne = std::numeric_limits&lt;double&gt;::quiet_NaN();
 237 :       current_statement__ = 4;
 238 :       nuMinusOne = in__.template read_constrain_lb&lt;local_scalar_t__,
 239 :                      jacobian__&gt;(0, lp__);
 240 :       double nu = std::numeric_limits&lt;double&gt;::quiet_NaN();
 241 :       out__.write(beta0);
 242 :       out__.write(beta1);
 243 :       out__.write(sigma);
 244 :       out__.write(nuMinusOne);
 245 :       if (stan::math::logical_negation(
 246 :             (stan::math::primitive_value(emit_transformed_parameters__) ||
 247 :             stan::math::primitive_value(emit_generated_quantities__)))) {
 248 :         return ;
 249 :       }
 250 :       current_statement__ = 6;
 251 :       nu = (nuMinusOne + 1);
 252 :       current_statement__ = 5;
 253 :       stan::math::check_greater_or_equal(function__, "nu", nu, 0);
 254 :       if (emit_transformed_parameters__) {
 255 :         out__.write(nu);
 256 :       }
 257 :       if (stan::math::logical_negation(emit_generated_quantities__)) {
 258 :         return ;
 259 :       }
 260 :     } catch (const std::exception&amp; e) {
 261 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 262 :     }
 263 :   }
 264 :   template &lt;typename VecVar, typename VecI,
 265 :             stan::require_vector_t&lt;VecVar&gt;* = nullptr,
 266 :             stan::require_vector_like_vt&lt;std::is_integral, VecI&gt;* = nullptr&gt;
 267 :   inline void
 268 :   unconstrain_array_impl(const VecVar&amp; params_r__, const VecI&amp; params_i__,
 269 :                          VecVar&amp; vars__, std::ostream* pstream__ = nullptr) const {
 270 :     using local_scalar_t__ = double;
 271 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 272 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 273 :     int current_statement__ = 0;
 274 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 275 :     // suppress unused var warning
 276 :     (void) DUMMY_VAR__;
 277 :     try {
 278 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
 279 :       pos__ = 1;
 280 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 281 :       current_statement__ = 1;
 282 :       beta0 = in__.read&lt;local_scalar_t__&gt;();
 283 :       out__.write(beta0);
 284 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 285 :       current_statement__ = 2;
 286 :       beta1 = in__.read&lt;local_scalar_t__&gt;();
 287 :       out__.write(beta1);
 288 :       local_scalar_t__ sigma = DUMMY_VAR__;
 289 :       current_statement__ = 3;
 290 :       sigma = in__.read&lt;local_scalar_t__&gt;();
 291 :       out__.write_free_lb(0, sigma);
 292 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 293 :       current_statement__ = 4;
 294 :       nuMinusOne = in__.read&lt;local_scalar_t__&gt;();
 295 :       out__.write_free_lb(0, nuMinusOne);
 296 :     } catch (const std::exception&amp; e) {
 297 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 298 :     }
 299 :   }
 300 :   template &lt;typename VecVar, stan::require_vector_t&lt;VecVar&gt;* = nullptr&gt;
 301 :   inline void
 302 :   transform_inits_impl(const stan::io::var_context&amp; context__, VecVar&amp;
 303 :                        vars__, std::ostream* pstream__ = nullptr) const {
 304 :     using local_scalar_t__ = double;
 305 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 306 :     int current_statement__ = 0;
 307 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 308 :     // suppress unused var warning
 309 :     (void) DUMMY_VAR__;
 310 :     try {
 311 :       current_statement__ = 1;
 312 :       context__.validate_dims("parameter initialization", "beta0", "double",
 313 :         std::vector&lt;size_t&gt;{});
 314 :       current_statement__ = 2;
 315 :       context__.validate_dims("parameter initialization", "beta1", "double",
 316 :         std::vector&lt;size_t&gt;{});
 317 :       current_statement__ = 3;
 318 :       context__.validate_dims("parameter initialization", "sigma", "double",
 319 :         std::vector&lt;size_t&gt;{});
 320 :       current_statement__ = 4;
 321 :       context__.validate_dims("parameter initialization", "nuMinusOne",
 322 :         "double", std::vector&lt;size_t&gt;{});
 323 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
 324 :       pos__ = 1;
 325 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 326 :       current_statement__ = 1;
 327 :       beta0 = context__.vals_r("beta0")[(1 - 1)];
 328 :       out__.write(beta0);
 329 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 330 :       current_statement__ = 2;
 331 :       beta1 = context__.vals_r("beta1")[(1 - 1)];
 332 :       out__.write(beta1);
 333 :       local_scalar_t__ sigma = DUMMY_VAR__;
 334 :       current_statement__ = 3;
 335 :       sigma = context__.vals_r("sigma")[(1 - 1)];
 336 :       out__.write_free_lb(0, sigma);
 337 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 338 :       current_statement__ = 4;
 339 :       nuMinusOne = context__.vals_r("nuMinusOne")[(1 - 1)];
 340 :       out__.write_free_lb(0, nuMinusOne);
 341 :     } catch (const std::exception&amp; e) {
 342 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 343 :     }
 344 :   }
 345 :   inline void
 346 :   get_param_names(std::vector&lt;std::string&gt;&amp; names__, const bool
 347 :                   emit_transformed_parameters__ = true, const bool
 348 :                   emit_generated_quantities__ = true) const {
 349 :     names__ = std::vector&lt;std::string&gt;{"beta0", "beta1", "sigma",
 350 :                 "nuMinusOne"};
 351 :     if (emit_transformed_parameters__) {
 352 :       std::vector&lt;std::string&gt; temp{"nu"};
 353 :       names__.reserve(names__.size() + temp.size());
 354 :       names__.insert(names__.end(), temp.begin(), temp.end());
 355 :     }
 356 :     if (emit_generated_quantities__) {}
 357 :   }
 358 :   inline void
 359 :   get_dims(std::vector&lt;std::vector&lt;size_t&gt;&gt;&amp; dimss__, const bool
 360 :            emit_transformed_parameters__ = true, const bool
 361 :            emit_generated_quantities__ = true) const {
 362 :     dimss__ = std::vector&lt;std::vector&lt;size_t&gt;&gt;{std::vector&lt;size_t&gt;{},
 363 :                 std::vector&lt;size_t&gt;{}, std::vector&lt;size_t&gt;{},
 364 :                 std::vector&lt;size_t&gt;{}};
 365 :     if (emit_transformed_parameters__) {
 366 :       std::vector&lt;std::vector&lt;size_t&gt;&gt; temp{std::vector&lt;size_t&gt;{}};
 367 :       dimss__.reserve(dimss__.size() + temp.size());
 368 :       dimss__.insert(dimss__.end(), temp.begin(), temp.end());
 369 :     }
 370 :     if (emit_generated_quantities__) {}
 371 :   }
 372 :   inline void
 373 :   constrained_param_names(std::vector&lt;std::string&gt;&amp; param_names__, bool
 374 :                           emit_transformed_parameters__ = true, bool
 375 :                           emit_generated_quantities__ = true) const final {
 376 :     param_names__.emplace_back(std::string() + "beta0");
 377 :     param_names__.emplace_back(std::string() + "beta1");
 378 :     param_names__.emplace_back(std::string() + "sigma");
 379 :     param_names__.emplace_back(std::string() + "nuMinusOne");
 380 :     if (emit_transformed_parameters__) {
 381 :       param_names__.emplace_back(std::string() + "nu");
 382 :     }
 383 :     if (emit_generated_quantities__) {}
 384 :   }
 385 :   inline void
 386 :   unconstrained_param_names(std::vector&lt;std::string&gt;&amp; param_names__, bool
 387 :                             emit_transformed_parameters__ = true, bool
 388 :                             emit_generated_quantities__ = true) const final {
 389 :     param_names__.emplace_back(std::string() + "beta0");
 390 :     param_names__.emplace_back(std::string() + "beta1");
 391 :     param_names__.emplace_back(std::string() + "sigma");
 392 :     param_names__.emplace_back(std::string() + "nuMinusOne");
 393 :     if (emit_transformed_parameters__) {
 394 :       param_names__.emplace_back(std::string() + "nu");
 395 :     }
 396 :     if (emit_generated_quantities__) {}
 397 :   }
 398 :   inline std::string get_constrained_sizedtypes() const {
 399 :     return std::string("[{\"name\":\"beta0\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta1\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"sigma\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nuMinusOne\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"}]");
 400 :   }
 401 :   inline std::string get_unconstrained_sizedtypes() const {
 402 :     return std::string("[{\"name\":\"beta0\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta1\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"sigma\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nuMinusOne\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"}]");
 403 :   }
 404 :   // Begin method overload boilerplate
 405 :   template &lt;typename RNG&gt; inline void
 406 :   write_array(RNG&amp; base_rng, Eigen::Matrix&lt;double,-1,1&gt;&amp; params_r,
 407 :               Eigen::Matrix&lt;double,-1,1&gt;&amp; vars, const bool
 408 :               emit_transformed_parameters = true, const bool
 409 :               emit_generated_quantities = true, std::ostream*
 410 :               pstream = nullptr) const {
 411 :     const size_t num_params__ = (((1 + 1) + 1) + 1);
 412 :     const size_t num_transformed = emit_transformed_parameters * (1);
 413 :     const size_t num_gen_quantities = emit_generated_quantities * (0);
 414 :     const size_t num_to_write = num_params__ + num_transformed +
 415 :       num_gen_quantities;
 416 :     std::vector&lt;int&gt; params_i;
 417 :     vars = Eigen::Matrix&lt;double,-1,1&gt;::Constant(num_to_write,
 418 :              std::numeric_limits&lt;double&gt;::quiet_NaN());
 419 :     write_array_impl(base_rng, params_r, params_i, vars,
 420 :       emit_transformed_parameters, emit_generated_quantities, pstream);
 421 :   }
 422 :   template &lt;typename RNG&gt; inline void
 423 :   write_array(RNG&amp; base_rng, std::vector&lt;double&gt;&amp; params_r, std::vector&lt;int&gt;&amp;
 424 :               params_i, std::vector&lt;double&gt;&amp; vars, bool
 425 :               emit_transformed_parameters = true, bool
 426 :               emit_generated_quantities = true, std::ostream*
 427 :               pstream = nullptr) const {
 428 :     const size_t num_params__ = (((1 + 1) + 1) + 1);
 429 :     const size_t num_transformed = emit_transformed_parameters * (1);
 430 :     const size_t num_gen_quantities = emit_generated_quantities * (0);
 431 :     const size_t num_to_write = num_params__ + num_transformed +
 432 :       num_gen_quantities;
 433 :     vars = std::vector&lt;double&gt;(num_to_write,
 434 :              std::numeric_limits&lt;double&gt;::quiet_NaN());
 435 :     write_array_impl(base_rng, params_r, params_i, vars,
 436 :       emit_transformed_parameters, emit_generated_quantities, pstream);
 437 :   }
 438 :   template &lt;bool propto__, bool jacobian__, typename T_&gt; inline T_
 439 :   log_prob(Eigen::Matrix&lt;T_,-1,1&gt;&amp; params_r, std::ostream* pstream = nullptr) const {
 440 :     Eigen::Matrix&lt;int,-1,1&gt; params_i;
 441 :     return log_prob_impl&lt;propto__, jacobian__&gt;(params_r, params_i, pstream);
 442 :   }
 443 :   template &lt;bool propto__, bool jacobian__, typename T_&gt; inline T_
 444 :   log_prob(std::vector&lt;T_&gt;&amp; params_r, std::vector&lt;int&gt;&amp; params_i,
 445 :            std::ostream* pstream = nullptr) const {
 446 :     return log_prob_impl&lt;propto__, jacobian__&gt;(params_r, params_i, pstream);
 447 :   }
 448 :   inline void
 449 :   transform_inits(const stan::io::var_context&amp; context,
 450 :                   Eigen::Matrix&lt;double,-1,1&gt;&amp; params_r, std::ostream*
 451 :                   pstream = nullptr) const final {
 452 :     std::vector&lt;double&gt; params_r_vec(params_r.size());
 453 :     std::vector&lt;int&gt; params_i;
 454 :     transform_inits(context, params_i, params_r_vec, pstream);
 455 :     params_r = Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(params_r_vec.data(),
 456 :                  params_r_vec.size());
 457 :   }
 458 :   inline void
 459 :   transform_inits(const stan::io::var_context&amp; context, std::vector&lt;int&gt;&amp;
 460 :                   params_i, std::vector&lt;double&gt;&amp; vars, std::ostream*
 461 :                   pstream__ = nullptr) const {
 462 :     vars.resize(num_params_r__);
 463 :     transform_inits_impl(context, vars, pstream__);
 464 :   }
 465 :   inline void
 466 :   unconstrain_array(const std::vector&lt;double&gt;&amp; params_constrained,
 467 :                     std::vector&lt;double&gt;&amp; params_unconstrained, std::ostream*
 468 :                     pstream = nullptr) const {
 469 :     const std::vector&lt;int&gt; params_i;
 470 :     params_unconstrained = std::vector&lt;double&gt;(num_params_r__,
 471 :                              std::numeric_limits&lt;double&gt;::quiet_NaN());
 472 :     unconstrain_array_impl(params_constrained, params_i,
 473 :       params_unconstrained, pstream);
 474 :   }
 475 :   inline void
 476 :   unconstrain_array(const Eigen::Matrix&lt;double,-1,1&gt;&amp; params_constrained,
 477 :                     Eigen::Matrix&lt;double,-1,1&gt;&amp; params_unconstrained,
 478 :                     std::ostream* pstream = nullptr) const {
 479 :     const std::vector&lt;int&gt; params_i;
 480 :     params_unconstrained = Eigen::Matrix&lt;double,-1,1&gt;::Constant(num_params_r__,
 481 :                              std::numeric_limits&lt;double&gt;::quiet_NaN());
 482 :     unconstrain_array_impl(params_constrained, params_i,
 483 :       params_unconstrained, pstream);
 484 :   }
 485 : };
 486 : }
 487 : using stan_model = model5ad42f2ccaea__namespace::model5ad42f2ccaea_;
 488 : #ifndef USING_R
 489 : // Boilerplate
 490 : stan::model::model_base&amp;
 491 : new_model(stan::io::var_context&amp; data_context, unsigned int seed,
 492 :           std::ostream* msg_stream) {
 493 :   stan_model* m = new stan_model(data_context, seed, msg_stream);
 494 :   return *m;
 495 : }
 496 : stan::math::profile_map&amp; get_stan_profile_data() {
 497 :   return model5ad42f2ccaea__namespace::profiles__;
 498 : }
 499 : #endif
 500 : #endif
 501 : 
 502 : RCPP_MODULE(stan_fit4model5ad42f2ccaea__mod) {
 503 :   class_&lt;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt; &gt;(
 504 :       "stan_fit4model5ad42f2ccaea_")
 505 : 
 506 :       .constructor&lt;SEXP, SEXP, SEXP&gt;()
 507 : 
 508 :       .method(
 509 :           "call_sampler",
 510 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::call_sampler)
 511 :       .method(
 512 :           "param_names",
 513 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::param_names)
 514 :       .method("param_names_oi",
 515 :               &amp;rstan::stan_fit&lt;stan_model,
 516 :                                boost::random::ecuyer1988&gt;::param_names_oi)
 517 :       .method("param_fnames_oi",
 518 :               &amp;rstan::stan_fit&lt;stan_model,
 519 :                                boost::random::ecuyer1988&gt;::param_fnames_oi)
 520 :       .method(
 521 :           "param_dims",
 522 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::param_dims)
 523 :       .method("param_dims_oi",
 524 :               &amp;rstan::stan_fit&lt;stan_model,
 525 :                                boost::random::ecuyer1988&gt;::param_dims_oi)
 526 :       .method("update_param_oi",
 527 :               &amp;rstan::stan_fit&lt;stan_model,
 528 :                                boost::random::ecuyer1988&gt;::update_param_oi)
 529 :       .method("param_oi_tidx",
 530 :               &amp;rstan::stan_fit&lt;stan_model,
 531 :                                boost::random::ecuyer1988&gt;::param_oi_tidx)
 532 :       .method("grad_log_prob",
 533 :               &amp;rstan::stan_fit&lt;stan_model,
 534 :                                boost::random::ecuyer1988&gt;::grad_log_prob)
 535 :       .method("log_prob",
 536 :               &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::log_prob)
 537 :       .method("unconstrain_pars",
 538 :               &amp;rstan::stan_fit&lt;stan_model,
 539 :                                boost::random::ecuyer1988&gt;::unconstrain_pars)
 540 :       .method("constrain_pars",
 541 :               &amp;rstan::stan_fit&lt;stan_model,
 542 :                                boost::random::ecuyer1988&gt;::constrain_pars)
 543 :       .method(
 544 :           "num_pars_unconstrained",
 545 :           &amp;rstan::stan_fit&lt;stan_model,
 546 :                            boost::random::ecuyer1988&gt;::num_pars_unconstrained)
 547 :       .method(
 548 :           "unconstrained_param_names",
 549 :           &amp;rstan::stan_fit&lt;
 550 :               stan_model, boost::random::ecuyer1988&gt;::unconstrained_param_names)
 551 :       .method(
 552 :           "constrained_param_names",
 553 :           &amp;rstan::stan_fit&lt;stan_model,
 554 :                            boost::random::ecuyer1988&gt;::constrained_param_names)
 555 :       .method("standalone_gqs",
 556 :               &amp;rstan::stan_fit&lt;stan_model,
 557 :                                boost::random::ecuyer1988&gt;::standalone_gqs);
 558 : }
 559 : 
 560 : 
 561 : // declarations
 562 : extern "C" {
 563 : SEXP file5ad440df72f5( ) ;
 564 : }
 565 : 
 566 : // definition
 567 : SEXP file5ad440df72f5() {
 568 :  return Rcpp::wrap("anon_model");
 569 : }</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>modelo_galton</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>S4 class stanmodel 'anon_model' coded as follows:
//
// This Stan program defines a simple model, with a
// vector of values 'y' modeled as normally distributed
// with mean 'mu' and standard deviation 'sigma'.
//
// Learn more about model development with Stan at:
//
//    http://mc-stan.org/users/interfaces/rstan.html
//    https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started
//
// The input data is a vector 'y' of length 'N'.
data {
  int&lt;lower=0&gt; N;     // N es un entero no negativo
  vector[N] y;          // y es un vector de longitud N de números reales
  vector[N] x;          // x es un vector de longitud N de números reales
}
// The parameters accepted by the model. Our model
// accepts two parameters 'mu' and 'sigma'.
parameters {
  real beta0;
  real beta1;
  real&lt;lower=0&gt; sigma;
  real&lt;lower=0&gt; nuMinusOne;
}
transformed parameters{
  real&lt;lower=0&gt; nu;
  nu = nuMinusOne + 1;
}
// The model to be estimated. We model the output
// 'y' to be normally distributed with mean 'mu'
// and standard deviation 'sigma'.
model {
  y ~ student_t(nu, beta0 + beta1 * x, sigma);
  beta0 ~ normal(0, 100);
  beta1 ~ normal(0, 4);
  sigma ~ uniform(6.0 / 100.0, 6.0 * 100.0);
  nuMinusOne ~ exponential(1/29.0);
} </code></pre>
</div>
</div>
<p>Y ajustamos el modelo:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstan)</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>galton_stanfit <span class="ot">&lt;-</span></span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sampling</span>(</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>    modelo_galton,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">N =</span> <span class="fu">nrow</span>(GaltonW),</span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> GaltonW<span class="sc">$</span>midparent,</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> GaltonW<span class="sc">$</span>height</span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">chains =</span> <span class="dv">4</span>,</span>
<span id="cb135-11"><a href="#cb135-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">iter =</span> <span class="dv">2000</span>,</span>
<span id="cb135-12"><a href="#cb135-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">warmup =</span> <span class="dv">1000</span></span>
<span id="cb135-13"><a href="#cb135-13" aria-hidden="true" tabindex="-1"></a>  )  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Ten en cuenta que los parámetros de pendiente e intercepción siguen estando correlacionados en el posterior, pero esto no molesta a Stan de la manera en que molesta a JAGS.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>galton_stanfit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Stan model: anon_model.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

              mean se_mean    sd    2.5%     25%     50%     75%   97.5% n_eff
beta0        17.19    0.16  5.87    5.99   13.27   17.18   21.12   28.68  1392
beta1         0.70    0.00  0.09    0.53    0.64    0.70    0.76    0.87  1388
sigma         1.94    0.00  0.13    1.69    1.86    1.94    2.03    2.19  1828
nuMinusOne   36.37    0.64 28.76    6.47   15.99   27.81   48.33  111.69  1995
nu           37.37    0.64 28.76    7.47   16.99   28.81   49.33  112.69  1995
lp__       -257.49    0.04  1.43 -261.22 -258.15 -257.16 -256.46 -255.75  1285
           Rhat
beta0         1
beta1         1
sigma         1
nuMinusOne    1
nu            1
lp__          1

Samples were drawn using NUTS(diag_e) at Mon Jun 17 22:08:29 2024.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(beta1 <span class="sc">~</span> beta0, <span class="at">data =</span> <span class="fu">posterior</span>(galton_stanfit), <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc.list</span>(galton_stanfit), </span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"beta0"</span>, <span class="st">"beta1"</span>, <span class="st">"sigma"</span>, <span class="st">"nu"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="modelo-con-dos-interceptos" class="level3" data-number="6.2.5">
<h3 data-number="6.2.5" class="anchored" data-anchor-id="modelo-con-dos-interceptos"><span class="header-section-number">6.2.5</span> Modelo con dos interceptos</h3>
<p>En el ejemplo anterior, hemos trabajado solo con las mujeres, pero podemos construir un modelo que maneje hombres y mujeres al mismo tiempo. Uno de estos modelos es el modelo de “múltiples interceptos”. En este modelo, ambos grupos (hombres y mujeres) tendrán la misma pendiente, pero los interceptos pueden diferir.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstan)</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">12345</span>)</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>modelo_galton2 <span class="ot">&lt;-</span> <span class="fu">stan_model</span>(<span class="at">file =</span> <span class="st">'galton2.stan'</span>,<span class="at">verbose =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
TRANSLATING MODEL '' FROM Stan CODE TO C++ CODE NOW.
OS: x86_64, linux-gnu; rstan: 2.32.5; Rcpp: 1.0.12; inline: 0.3.19 
 &gt;&gt; setting environment variables: 
PKG_LIBS =  '/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/rstan/lib//libStanServices.a' -L'/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/lib/' -lStanHeaders -L'/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppParallel/lib/' -ltbb 
PKG_CPPFLAGS =   -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/Rcpp/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppEigen/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppEigen/include/unsupported"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/BH/include" -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/src/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/RcppParallel/include/"  -I"/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/rstan/include" -DEIGEN_NO_DEBUG  -DBOOST_DISABLE_ASSERTS  -DBOOST_PENDING_INTEGER_LOG2_HPP  -DSTAN_THREADS  -DUSE_STANC3 -DSTRICT_R_HEADERS  -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION  -D_HAS_AUTO_PTR_ETC=0  -include '/home/luis/R/x86_64-redhat-linux-gnu-library/4.3/StanHeaders/include/stan/math/prim/fun/Eigen.hpp'  -D_REENTRANT -DRCPP_PARALLEL_USE_TBB=1 
 &gt;&gt; Program source :

   1 : 
   2 : // includes from the plugin
   3 : // [[Rcpp::plugins(cpp14)]]
   4 : 
   5 : 
   6 : // user includes
   7 : #include &lt;Rcpp.h&gt;
   8 : using namespace Rcpp;
   9 : #ifndef MODELS_HPP
  10 : #define MODELS_HPP
  11 : #define STAN__SERVICES__COMMAND_HPP
  12 : #include &lt;rstan/rstaninc.hpp&gt;
  13 : #ifndef USE_STANC3
  14 : #define USE_STANC3
  15 : #endif
  16 : // Code generated by stanc v2.32.2
  17 : #include &lt;stan/model/model_header.hpp&gt;
  18 : namespace model5ad477487001__namespace {
  19 : using stan::model::model_base_crtp;
  20 : using namespace stan::math;
  21 : stan::math::profile_map profiles__;
  22 : static constexpr std::array&lt;const char*, 25&gt; locations_array__ =
  23 :   {" (found before start of program)",
  24 :   " (in 'anon_model', line 21, column 2 to column 13)",
  25 :   " (in 'anon_model', line 22, column 2 to column 13)",
  26 :   " (in 'anon_model', line 23, column 2 to column 13)",
  27 :   " (in 'anon_model', line 24, column 2 to column 22)",
  28 :   " (in 'anon_model', line 25, column 2 to column 27)",
  29 :   " (in 'anon_model', line 28, column 2 to column 24)",
  30 :   " (in 'anon_model', line 29, column 2 to column 19)",
  31 :   " (in 'anon_model', line 30, column 2 to column 22)",
  32 :   " (in 'anon_model', line 31, column 2 to column 22)",
  33 :   " (in 'anon_model', line 38, column 4 to column 69)",
  34 :   " (in 'anon_model', line 37, column 17 to line 39, column 3)",
  35 :   " (in 'anon_model', line 37, column 2 to line 39, column 3)",
  36 :   " (in 'anon_model', line 40, column 2 to column 25)",
  37 :   " (in 'anon_model', line 41, column 2 to column 23)",
  38 :   " (in 'anon_model', line 42, column 2 to column 23)",
  39 :   " (in 'anon_model', line 43, column 2 to column 44)",
  40 :   " (in 'anon_model', line 44, column 2 to column 35)",
  41 :   " (in 'anon_model', line 13, column 2 to column 17)",
  42 :   " (in 'anon_model', line 14, column 9 to column 10)",
  43 :   " (in 'anon_model', line 14, column 2 to column 14)",
  44 :   " (in 'anon_model', line 15, column 9 to column 10)",
  45 :   " (in 'anon_model', line 15, column 2 to column 14)",
  46 :   " (in 'anon_model', line 16, column 17 to column 18)",
  47 :   " (in 'anon_model', line 16, column 2 to column 20)"};
  48 : class model5ad477487001_ final : public model_base_crtp&lt;model5ad477487001_&gt; {
  49 : private:
  50 :   int N;
  51 :   Eigen::Matrix&lt;double,-1,1&gt; y_data__;
  52 :   Eigen::Matrix&lt;double,-1,1&gt; x_data__;
  53 :   std::vector&lt;int&gt; g;
  54 :   Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt; y{nullptr, 0};
  55 :   Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt; x{nullptr, 0};
  56 : public:
  57 :   ~model5ad477487001_() {}
  58 :   model5ad477487001_(stan::io::var_context&amp; context__, unsigned int
  59 :                      random_seed__ = 0, std::ostream* pstream__ = nullptr)
  60 :       : model_base_crtp(0) {
  61 :     int current_statement__ = 0;
  62 :     using local_scalar_t__ = double;
  63 :     boost::ecuyer1988 base_rng__ =
  64 :       stan::services::util::create_rng(random_seed__, 0);
  65 :     // suppress unused var warning
  66 :     (void) base_rng__;
  67 :     static constexpr const char* function__ =
  68 :       "model5ad477487001__namespace::model5ad477487001_";
  69 :     // suppress unused var warning
  70 :     (void) function__;
  71 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
  72 :     // suppress unused var warning
  73 :     (void) DUMMY_VAR__;
  74 :     try {
  75 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
  76 :       pos__ = 1;
  77 :       current_statement__ = 18;
  78 :       context__.validate_dims("data initialization", "N", "int",
  79 :         std::vector&lt;size_t&gt;{});
  80 :       N = std::numeric_limits&lt;int&gt;::min();
  81 :       current_statement__ = 18;
  82 :       N = context__.vals_i("N")[(1 - 1)];
  83 :       current_statement__ = 18;
  84 :       stan::math::check_greater_or_equal(function__, "N", N, 0);
  85 :       current_statement__ = 19;
  86 :       stan::math::validate_non_negative_index("y", "N", N);
  87 :       current_statement__ = 20;
  88 :       context__.validate_dims("data initialization", "y", "double",
  89 :         std::vector&lt;size_t&gt;{static_cast&lt;size_t&gt;(N)});
  90 :       y_data__ = Eigen::Matrix&lt;double,-1,1&gt;::Constant(N,
  91 :                    std::numeric_limits&lt;double&gt;::quiet_NaN());
  92 :       new (&amp;y) Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(y_data__.data(), N);
  93 :       {
  94 :         std::vector&lt;local_scalar_t__&gt; y_flat__;
  95 :         current_statement__ = 20;
  96 :         y_flat__ = context__.vals_r("y");
  97 :         current_statement__ = 20;
  98 :         pos__ = 1;
  99 :         current_statement__ = 20;
 100 :         for (int sym1__ = 1; sym1__ &lt;= N; ++sym1__) {
 101 :           current_statement__ = 20;
 102 :           stan::model::assign(y, y_flat__[(pos__ - 1)],
 103 :             "assigning variable y", stan::model::index_uni(sym1__));
 104 :           current_statement__ = 20;
 105 :           pos__ = (pos__ + 1);
 106 :         }
 107 :       }
 108 :       current_statement__ = 21;
 109 :       stan::math::validate_non_negative_index("x", "N", N);
 110 :       current_statement__ = 22;
 111 :       context__.validate_dims("data initialization", "x", "double",
 112 :         std::vector&lt;size_t&gt;{static_cast&lt;size_t&gt;(N)});
 113 :       x_data__ = Eigen::Matrix&lt;double,-1,1&gt;::Constant(N,
 114 :                    std::numeric_limits&lt;double&gt;::quiet_NaN());
 115 :       new (&amp;x) Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(x_data__.data(), N);
 116 :       {
 117 :         std::vector&lt;local_scalar_t__&gt; x_flat__;
 118 :         current_statement__ = 22;
 119 :         x_flat__ = context__.vals_r("x");
 120 :         current_statement__ = 22;
 121 :         pos__ = 1;
 122 :         current_statement__ = 22;
 123 :         for (int sym1__ = 1; sym1__ &lt;= N; ++sym1__) {
 124 :           current_statement__ = 22;
 125 :           stan::model::assign(x, x_flat__[(pos__ - 1)],
 126 :             "assigning variable x", stan::model::index_uni(sym1__));
 127 :           current_statement__ = 22;
 128 :           pos__ = (pos__ + 1);
 129 :         }
 130 :       }
 131 :       current_statement__ = 23;
 132 :       stan::math::validate_non_negative_index("g", "N", N);
 133 :       current_statement__ = 24;
 134 :       context__.validate_dims("data initialization", "g", "int",
 135 :         std::vector&lt;size_t&gt;{static_cast&lt;size_t&gt;(N)});
 136 :       g = std::vector&lt;int&gt;(N, std::numeric_limits&lt;int&gt;::min());
 137 :       current_statement__ = 24;
 138 :       g = context__.vals_i("g");
 139 :       current_statement__ = 24;
 140 :       stan::math::check_greater_or_equal(function__, "g", g, 0);
 141 :     } catch (const std::exception&amp; e) {
 142 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 143 :     }
 144 :     num_params_r__ = 1 + 1 + 1 + 1 + 1;
 145 :   }
 146 :   inline std::string model_name() const final {
 147 :     return "model5ad477487001_";
 148 :   }
 149 :   inline std::vector&lt;std::string&gt; model_compile_info() const noexcept {
 150 :     return std::vector&lt;std::string&gt;{"stanc_version = stanc3 v2.32.2",
 151 :              "stancflags = --"};
 152 :   }
 153 :   template &lt;bool propto__, bool jacobian__, typename VecR, typename VecI,
 154 :             stan::require_vector_like_t&lt;VecR&gt;* = nullptr,
 155 :             stan::require_vector_like_vt&lt;std::is_integral, VecI&gt;* = nullptr&gt;
 156 :   inline stan::scalar_type_t&lt;VecR&gt;
 157 :   log_prob_impl(VecR&amp; params_r__, VecI&amp; params_i__, std::ostream*
 158 :                 pstream__ = nullptr) const {
 159 :     using T__ = stan::scalar_type_t&lt;VecR&gt;;
 160 :     using local_scalar_t__ = T__;
 161 :     T__ lp__(0.0);
 162 :     stan::math::accumulator&lt;T__&gt; lp_accum__;
 163 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 164 :     int current_statement__ = 0;
 165 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 166 :     // suppress unused var warning
 167 :     (void) DUMMY_VAR__;
 168 :     static constexpr const char* function__ =
 169 :       "model5ad477487001__namespace::log_prob";
 170 :     // suppress unused var warning
 171 :     (void) function__;
 172 :     try {
 173 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 174 :       current_statement__ = 1;
 175 :       beta0 = in__.template read&lt;local_scalar_t__&gt;();
 176 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 177 :       current_statement__ = 2;
 178 :       beta1 = in__.template read&lt;local_scalar_t__&gt;();
 179 :       local_scalar_t__ beta2 = DUMMY_VAR__;
 180 :       current_statement__ = 3;
 181 :       beta2 = in__.template read&lt;local_scalar_t__&gt;();
 182 :       local_scalar_t__ sigma = DUMMY_VAR__;
 183 :       current_statement__ = 4;
 184 :       sigma = in__.template read_constrain_lb&lt;local_scalar_t__,
 185 :                 jacobian__&gt;(0, lp__);
 186 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 187 :       current_statement__ = 5;
 188 :       nuMinusOne = in__.template read_constrain_lb&lt;local_scalar_t__,
 189 :                      jacobian__&gt;(0, lp__);
 190 :       local_scalar_t__ log10nu = DUMMY_VAR__;
 191 :       local_scalar_t__ nu = DUMMY_VAR__;
 192 :       current_statement__ = 8;
 193 :       nu = (nuMinusOne + 1);
 194 :       current_statement__ = 9;
 195 :       log10nu = stan::math::log10(nu);
 196 :       current_statement__ = 6;
 197 :       stan::math::check_greater_or_equal(function__, "log10nu", log10nu, 0);
 198 :       current_statement__ = 7;
 199 :       stan::math::check_greater_or_equal(function__, "nu", nu, 0);
 200 :       {
 201 :         current_statement__ = 12;
 202 :         for (int i = 1; i &lt;= N; ++i) {
 203 :           current_statement__ = 10;
 204 :           lp_accum__.add(stan::math::student_t_lpdf&lt;propto__&gt;(
 205 :                            stan::model::rvalue(y, "y",
 206 :                              stan::model::index_uni(i)), nu, ((beta0 + (beta1
 207 :                            *
 208 :                            stan::model::rvalue(x, "x",
 209 :                              stan::model::index_uni(i)))) + (beta2 *
 210 :                            stan::model::rvalue(g, "g",
 211 :                              stan::model::index_uni(i)))), sigma));
 212 :         }
 213 :         current_statement__ = 13;
 214 :         lp_accum__.add(stan::math::normal_lpdf&lt;propto__&gt;(beta0, 0, 100));
 215 :         current_statement__ = 14;
 216 :         lp_accum__.add(stan::math::normal_lpdf&lt;propto__&gt;(beta1, 0, 4));
 217 :         current_statement__ = 15;
 218 :         lp_accum__.add(stan::math::normal_lpdf&lt;propto__&gt;(beta2, 0, 4));
 219 :         current_statement__ = 16;
 220 :         lp_accum__.add(stan::math::uniform_lpdf&lt;propto__&gt;(sigma, (6.0 /
 221 :                          100.0), (6.0 * 100.0)));
 222 :         current_statement__ = 17;
 223 :         lp_accum__.add(stan::math::exponential_lpdf&lt;propto__&gt;(nuMinusOne, (1
 224 :                          / 29.0)));
 225 :       }
 226 :     } catch (const std::exception&amp; e) {
 227 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 228 :     }
 229 :     lp_accum__.add(lp__);
 230 :     return lp_accum__.sum();
 231 :   }
 232 :   template &lt;typename RNG, typename VecR, typename VecI, typename VecVar,
 233 :             stan::require_vector_like_vt&lt;std::is_floating_point,
 234 :             VecR&gt;* = nullptr, stan::require_vector_like_vt&lt;std::is_integral,
 235 :             VecI&gt;* = nullptr, stan::require_vector_vt&lt;std::is_floating_point,
 236 :             VecVar&gt;* = nullptr&gt;
 237 :   inline void
 238 :   write_array_impl(RNG&amp; base_rng__, VecR&amp; params_r__, VecI&amp; params_i__,
 239 :                    VecVar&amp; vars__, const bool
 240 :                    emit_transformed_parameters__ = true, const bool
 241 :                    emit_generated_quantities__ = true, std::ostream*
 242 :                    pstream__ = nullptr) const {
 243 :     using local_scalar_t__ = double;
 244 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 245 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 246 :     static constexpr bool propto__ = true;
 247 :     // suppress unused var warning
 248 :     (void) propto__;
 249 :     double lp__ = 0.0;
 250 :     // suppress unused var warning
 251 :     (void) lp__;
 252 :     int current_statement__ = 0;
 253 :     stan::math::accumulator&lt;double&gt; lp_accum__;
 254 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 255 :     // suppress unused var warning
 256 :     (void) DUMMY_VAR__;
 257 :     constexpr bool jacobian__ = false;
 258 :     static constexpr const char* function__ =
 259 :       "model5ad477487001__namespace::write_array";
 260 :     // suppress unused var warning
 261 :     (void) function__;
 262 :     try {
 263 :       double beta0 = std::numeric_limits&lt;double&gt;::quiet_NaN();
 264 :       current_statement__ = 1;
 265 :       beta0 = in__.template read&lt;local_scalar_t__&gt;();
 266 :       double beta1 = std::numeric_limits&lt;double&gt;::quiet_NaN();
 267 :       current_statement__ = 2;
 268 :       beta1 = in__.template read&lt;local_scalar_t__&gt;();
 269 :       double beta2 = std::numeric_limits&lt;double&gt;::quiet_NaN();
 270 :       current_statement__ = 3;
 271 :       beta2 = in__.template read&lt;local_scalar_t__&gt;();
 272 :       double sigma = std::numeric_limits&lt;double&gt;::quiet_NaN();
 273 :       current_statement__ = 4;
 274 :       sigma = in__.template read_constrain_lb&lt;local_scalar_t__,
 275 :                 jacobian__&gt;(0, lp__);
 276 :       double nuMinusOne = std::numeric_limits&lt;double&gt;::quiet_NaN();
 277 :       current_statement__ = 5;
 278 :       nuMinusOne = in__.template read_constrain_lb&lt;local_scalar_t__,
 279 :                      jacobian__&gt;(0, lp__);
 280 :       double log10nu = std::numeric_limits&lt;double&gt;::quiet_NaN();
 281 :       double nu = std::numeric_limits&lt;double&gt;::quiet_NaN();
 282 :       out__.write(beta0);
 283 :       out__.write(beta1);
 284 :       out__.write(beta2);
 285 :       out__.write(sigma);
 286 :       out__.write(nuMinusOne);
 287 :       if (stan::math::logical_negation(
 288 :             (stan::math::primitive_value(emit_transformed_parameters__) ||
 289 :             stan::math::primitive_value(emit_generated_quantities__)))) {
 290 :         return ;
 291 :       }
 292 :       current_statement__ = 8;
 293 :       nu = (nuMinusOne + 1);
 294 :       current_statement__ = 9;
 295 :       log10nu = stan::math::log10(nu);
 296 :       current_statement__ = 6;
 297 :       stan::math::check_greater_or_equal(function__, "log10nu", log10nu, 0);
 298 :       current_statement__ = 7;
 299 :       stan::math::check_greater_or_equal(function__, "nu", nu, 0);
 300 :       if (emit_transformed_parameters__) {
 301 :         out__.write(log10nu);
 302 :         out__.write(nu);
 303 :       }
 304 :       if (stan::math::logical_negation(emit_generated_quantities__)) {
 305 :         return ;
 306 :       }
 307 :     } catch (const std::exception&amp; e) {
 308 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 309 :     }
 310 :   }
 311 :   template &lt;typename VecVar, typename VecI,
 312 :             stan::require_vector_t&lt;VecVar&gt;* = nullptr,
 313 :             stan::require_vector_like_vt&lt;std::is_integral, VecI&gt;* = nullptr&gt;
 314 :   inline void
 315 :   unconstrain_array_impl(const VecVar&amp; params_r__, const VecI&amp; params_i__,
 316 :                          VecVar&amp; vars__, std::ostream* pstream__ = nullptr) const {
 317 :     using local_scalar_t__ = double;
 318 :     stan::io::deserializer&lt;local_scalar_t__&gt; in__(params_r__, params_i__);
 319 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 320 :     int current_statement__ = 0;
 321 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 322 :     // suppress unused var warning
 323 :     (void) DUMMY_VAR__;
 324 :     try {
 325 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
 326 :       pos__ = 1;
 327 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 328 :       current_statement__ = 1;
 329 :       beta0 = in__.read&lt;local_scalar_t__&gt;();
 330 :       out__.write(beta0);
 331 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 332 :       current_statement__ = 2;
 333 :       beta1 = in__.read&lt;local_scalar_t__&gt;();
 334 :       out__.write(beta1);
 335 :       local_scalar_t__ beta2 = DUMMY_VAR__;
 336 :       current_statement__ = 3;
 337 :       beta2 = in__.read&lt;local_scalar_t__&gt;();
 338 :       out__.write(beta2);
 339 :       local_scalar_t__ sigma = DUMMY_VAR__;
 340 :       current_statement__ = 4;
 341 :       sigma = in__.read&lt;local_scalar_t__&gt;();
 342 :       out__.write_free_lb(0, sigma);
 343 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 344 :       current_statement__ = 5;
 345 :       nuMinusOne = in__.read&lt;local_scalar_t__&gt;();
 346 :       out__.write_free_lb(0, nuMinusOne);
 347 :     } catch (const std::exception&amp; e) {
 348 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 349 :     }
 350 :   }
 351 :   template &lt;typename VecVar, stan::require_vector_t&lt;VecVar&gt;* = nullptr&gt;
 352 :   inline void
 353 :   transform_inits_impl(const stan::io::var_context&amp; context__, VecVar&amp;
 354 :                        vars__, std::ostream* pstream__ = nullptr) const {
 355 :     using local_scalar_t__ = double;
 356 :     stan::io::serializer&lt;local_scalar_t__&gt; out__(vars__);
 357 :     int current_statement__ = 0;
 358 :     local_scalar_t__ DUMMY_VAR__(std::numeric_limits&lt;double&gt;::quiet_NaN());
 359 :     // suppress unused var warning
 360 :     (void) DUMMY_VAR__;
 361 :     try {
 362 :       current_statement__ = 1;
 363 :       context__.validate_dims("parameter initialization", "beta0", "double",
 364 :         std::vector&lt;size_t&gt;{});
 365 :       current_statement__ = 2;
 366 :       context__.validate_dims("parameter initialization", "beta1", "double",
 367 :         std::vector&lt;size_t&gt;{});
 368 :       current_statement__ = 3;
 369 :       context__.validate_dims("parameter initialization", "beta2", "double",
 370 :         std::vector&lt;size_t&gt;{});
 371 :       current_statement__ = 4;
 372 :       context__.validate_dims("parameter initialization", "sigma", "double",
 373 :         std::vector&lt;size_t&gt;{});
 374 :       current_statement__ = 5;
 375 :       context__.validate_dims("parameter initialization", "nuMinusOne",
 376 :         "double", std::vector&lt;size_t&gt;{});
 377 :       int pos__ = std::numeric_limits&lt;int&gt;::min();
 378 :       pos__ = 1;
 379 :       local_scalar_t__ beta0 = DUMMY_VAR__;
 380 :       current_statement__ = 1;
 381 :       beta0 = context__.vals_r("beta0")[(1 - 1)];
 382 :       out__.write(beta0);
 383 :       local_scalar_t__ beta1 = DUMMY_VAR__;
 384 :       current_statement__ = 2;
 385 :       beta1 = context__.vals_r("beta1")[(1 - 1)];
 386 :       out__.write(beta1);
 387 :       local_scalar_t__ beta2 = DUMMY_VAR__;
 388 :       current_statement__ = 3;
 389 :       beta2 = context__.vals_r("beta2")[(1 - 1)];
 390 :       out__.write(beta2);
 391 :       local_scalar_t__ sigma = DUMMY_VAR__;
 392 :       current_statement__ = 4;
 393 :       sigma = context__.vals_r("sigma")[(1 - 1)];
 394 :       out__.write_free_lb(0, sigma);
 395 :       local_scalar_t__ nuMinusOne = DUMMY_VAR__;
 396 :       current_statement__ = 5;
 397 :       nuMinusOne = context__.vals_r("nuMinusOne")[(1 - 1)];
 398 :       out__.write_free_lb(0, nuMinusOne);
 399 :     } catch (const std::exception&amp; e) {
 400 :       stan::lang::rethrow_located(e, locations_array__[current_statement__]);
 401 :     }
 402 :   }
 403 :   inline void
 404 :   get_param_names(std::vector&lt;std::string&gt;&amp; names__, const bool
 405 :                   emit_transformed_parameters__ = true, const bool
 406 :                   emit_generated_quantities__ = true) const {
 407 :     names__ = std::vector&lt;std::string&gt;{"beta0", "beta1", "beta2", "sigma",
 408 :                 "nuMinusOne"};
 409 :     if (emit_transformed_parameters__) {
 410 :       std::vector&lt;std::string&gt; temp{"log10nu", "nu"};
 411 :       names__.reserve(names__.size() + temp.size());
 412 :       names__.insert(names__.end(), temp.begin(), temp.end());
 413 :     }
 414 :     if (emit_generated_quantities__) {}
 415 :   }
 416 :   inline void
 417 :   get_dims(std::vector&lt;std::vector&lt;size_t&gt;&gt;&amp; dimss__, const bool
 418 :            emit_transformed_parameters__ = true, const bool
 419 :            emit_generated_quantities__ = true) const {
 420 :     dimss__ = std::vector&lt;std::vector&lt;size_t&gt;&gt;{std::vector&lt;size_t&gt;{},
 421 :                 std::vector&lt;size_t&gt;{}, std::vector&lt;size_t&gt;{},
 422 :                 std::vector&lt;size_t&gt;{}, std::vector&lt;size_t&gt;{}};
 423 :     if (emit_transformed_parameters__) {
 424 :       std::vector&lt;std::vector&lt;size_t&gt;&gt;
 425 :         temp{std::vector&lt;size_t&gt;{}, std::vector&lt;size_t&gt;{}};
 426 :       dimss__.reserve(dimss__.size() + temp.size());
 427 :       dimss__.insert(dimss__.end(), temp.begin(), temp.end());
 428 :     }
 429 :     if (emit_generated_quantities__) {}
 430 :   }
 431 :   inline void
 432 :   constrained_param_names(std::vector&lt;std::string&gt;&amp; param_names__, bool
 433 :                           emit_transformed_parameters__ = true, bool
 434 :                           emit_generated_quantities__ = true) const final {
 435 :     param_names__.emplace_back(std::string() + "beta0");
 436 :     param_names__.emplace_back(std::string() + "beta1");
 437 :     param_names__.emplace_back(std::string() + "beta2");
 438 :     param_names__.emplace_back(std::string() + "sigma");
 439 :     param_names__.emplace_back(std::string() + "nuMinusOne");
 440 :     if (emit_transformed_parameters__) {
 441 :       param_names__.emplace_back(std::string() + "log10nu");
 442 :       param_names__.emplace_back(std::string() + "nu");
 443 :     }
 444 :     if (emit_generated_quantities__) {}
 445 :   }
 446 :   inline void
 447 :   unconstrained_param_names(std::vector&lt;std::string&gt;&amp; param_names__, bool
 448 :                             emit_transformed_parameters__ = true, bool
 449 :                             emit_generated_quantities__ = true) const final {
 450 :     param_names__.emplace_back(std::string() + "beta0");
 451 :     param_names__.emplace_back(std::string() + "beta1");
 452 :     param_names__.emplace_back(std::string() + "beta2");
 453 :     param_names__.emplace_back(std::string() + "sigma");
 454 :     param_names__.emplace_back(std::string() + "nuMinusOne");
 455 :     if (emit_transformed_parameters__) {
 456 :       param_names__.emplace_back(std::string() + "log10nu");
 457 :       param_names__.emplace_back(std::string() + "nu");
 458 :     }
 459 :     if (emit_generated_quantities__) {}
 460 :   }
 461 :   inline std::string get_constrained_sizedtypes() const {
 462 :     return std::string("[{\"name\":\"beta0\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta1\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta2\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"sigma\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nuMinusOne\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"log10nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"},{\"name\":\"nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"}]");
 463 :   }
 464 :   inline std::string get_unconstrained_sizedtypes() const {
 465 :     return std::string("[{\"name\":\"beta0\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta1\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"beta2\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"sigma\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"nuMinusOne\",\"type\":{\"name\":\"real\"},\"block\":\"parameters\"},{\"name\":\"log10nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"},{\"name\":\"nu\",\"type\":{\"name\":\"real\"},\"block\":\"transformed_parameters\"}]");
 466 :   }
 467 :   // Begin method overload boilerplate
 468 :   template &lt;typename RNG&gt; inline void
 469 :   write_array(RNG&amp; base_rng, Eigen::Matrix&lt;double,-1,1&gt;&amp; params_r,
 470 :               Eigen::Matrix&lt;double,-1,1&gt;&amp; vars, const bool
 471 :               emit_transformed_parameters = true, const bool
 472 :               emit_generated_quantities = true, std::ostream*
 473 :               pstream = nullptr) const {
 474 :     const size_t num_params__ = ((((1 + 1) + 1) + 1) + 1);
 475 :     const size_t num_transformed = emit_transformed_parameters * ((1 + 1));
 476 :     const size_t num_gen_quantities = emit_generated_quantities * (0);
 477 :     const size_t num_to_write = num_params__ + num_transformed +
 478 :       num_gen_quantities;
 479 :     std::vector&lt;int&gt; params_i;
 480 :     vars = Eigen::Matrix&lt;double,-1,1&gt;::Constant(num_to_write,
 481 :              std::numeric_limits&lt;double&gt;::quiet_NaN());
 482 :     write_array_impl(base_rng, params_r, params_i, vars,
 483 :       emit_transformed_parameters, emit_generated_quantities, pstream);
 484 :   }
 485 :   template &lt;typename RNG&gt; inline void
 486 :   write_array(RNG&amp; base_rng, std::vector&lt;double&gt;&amp; params_r, std::vector&lt;int&gt;&amp;
 487 :               params_i, std::vector&lt;double&gt;&amp; vars, bool
 488 :               emit_transformed_parameters = true, bool
 489 :               emit_generated_quantities = true, std::ostream*
 490 :               pstream = nullptr) const {
 491 :     const size_t num_params__ = ((((1 + 1) + 1) + 1) + 1);
 492 :     const size_t num_transformed = emit_transformed_parameters * ((1 + 1));
 493 :     const size_t num_gen_quantities = emit_generated_quantities * (0);
 494 :     const size_t num_to_write = num_params__ + num_transformed +
 495 :       num_gen_quantities;
 496 :     vars = std::vector&lt;double&gt;(num_to_write,
 497 :              std::numeric_limits&lt;double&gt;::quiet_NaN());
 498 :     write_array_impl(base_rng, params_r, params_i, vars,
 499 :       emit_transformed_parameters, emit_generated_quantities, pstream);
 500 :   }
 501 :   template &lt;bool propto__, bool jacobian__, typename T_&gt; inline T_
 502 :   log_prob(Eigen::Matrix&lt;T_,-1,1&gt;&amp; params_r, std::ostream* pstream = nullptr) const {
 503 :     Eigen::Matrix&lt;int,-1,1&gt; params_i;
 504 :     return log_prob_impl&lt;propto__, jacobian__&gt;(params_r, params_i, pstream);
 505 :   }
 506 :   template &lt;bool propto__, bool jacobian__, typename T_&gt; inline T_
 507 :   log_prob(std::vector&lt;T_&gt;&amp; params_r, std::vector&lt;int&gt;&amp; params_i,
 508 :            std::ostream* pstream = nullptr) const {
 509 :     return log_prob_impl&lt;propto__, jacobian__&gt;(params_r, params_i, pstream);
 510 :   }
 511 :   inline void
 512 :   transform_inits(const stan::io::var_context&amp; context,
 513 :                   Eigen::Matrix&lt;double,-1,1&gt;&amp; params_r, std::ostream*
 514 :                   pstream = nullptr) const final {
 515 :     std::vector&lt;double&gt; params_r_vec(params_r.size());
 516 :     std::vector&lt;int&gt; params_i;
 517 :     transform_inits(context, params_i, params_r_vec, pstream);
 518 :     params_r = Eigen::Map&lt;Eigen::Matrix&lt;double,-1,1&gt;&gt;(params_r_vec.data(),
 519 :                  params_r_vec.size());
 520 :   }
 521 :   inline void
 522 :   transform_inits(const stan::io::var_context&amp; context, std::vector&lt;int&gt;&amp;
 523 :                   params_i, std::vector&lt;double&gt;&amp; vars, std::ostream*
 524 :                   pstream__ = nullptr) const {
 525 :     vars.resize(num_params_r__);
 526 :     transform_inits_impl(context, vars, pstream__);
 527 :   }
 528 :   inline void
 529 :   unconstrain_array(const std::vector&lt;double&gt;&amp; params_constrained,
 530 :                     std::vector&lt;double&gt;&amp; params_unconstrained, std::ostream*
 531 :                     pstream = nullptr) const {
 532 :     const std::vector&lt;int&gt; params_i;
 533 :     params_unconstrained = std::vector&lt;double&gt;(num_params_r__,
 534 :                              std::numeric_limits&lt;double&gt;::quiet_NaN());
 535 :     unconstrain_array_impl(params_constrained, params_i,
 536 :       params_unconstrained, pstream);
 537 :   }
 538 :   inline void
 539 :   unconstrain_array(const Eigen::Matrix&lt;double,-1,1&gt;&amp; params_constrained,
 540 :                     Eigen::Matrix&lt;double,-1,1&gt;&amp; params_unconstrained,
 541 :                     std::ostream* pstream = nullptr) const {
 542 :     const std::vector&lt;int&gt; params_i;
 543 :     params_unconstrained = Eigen::Matrix&lt;double,-1,1&gt;::Constant(num_params_r__,
 544 :                              std::numeric_limits&lt;double&gt;::quiet_NaN());
 545 :     unconstrain_array_impl(params_constrained, params_i,
 546 :       params_unconstrained, pstream);
 547 :   }
 548 : };
 549 : }
 550 : using stan_model = model5ad477487001__namespace::model5ad477487001_;
 551 : #ifndef USING_R
 552 : // Boilerplate
 553 : stan::model::model_base&amp;
 554 : new_model(stan::io::var_context&amp; data_context, unsigned int seed,
 555 :           std::ostream* msg_stream) {
 556 :   stan_model* m = new stan_model(data_context, seed, msg_stream);
 557 :   return *m;
 558 : }
 559 : stan::math::profile_map&amp; get_stan_profile_data() {
 560 :   return model5ad477487001__namespace::profiles__;
 561 : }
 562 : #endif
 563 : #endif
 564 : 
 565 : RCPP_MODULE(stan_fit4model5ad477487001__mod) {
 566 :   class_&lt;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt; &gt;(
 567 :       "stan_fit4model5ad477487001_")
 568 : 
 569 :       .constructor&lt;SEXP, SEXP, SEXP&gt;()
 570 : 
 571 :       .method(
 572 :           "call_sampler",
 573 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::call_sampler)
 574 :       .method(
 575 :           "param_names",
 576 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::param_names)
 577 :       .method("param_names_oi",
 578 :               &amp;rstan::stan_fit&lt;stan_model,
 579 :                                boost::random::ecuyer1988&gt;::param_names_oi)
 580 :       .method("param_fnames_oi",
 581 :               &amp;rstan::stan_fit&lt;stan_model,
 582 :                                boost::random::ecuyer1988&gt;::param_fnames_oi)
 583 :       .method(
 584 :           "param_dims",
 585 :           &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::param_dims)
 586 :       .method("param_dims_oi",
 587 :               &amp;rstan::stan_fit&lt;stan_model,
 588 :                                boost::random::ecuyer1988&gt;::param_dims_oi)
 589 :       .method("update_param_oi",
 590 :               &amp;rstan::stan_fit&lt;stan_model,
 591 :                                boost::random::ecuyer1988&gt;::update_param_oi)
 592 :       .method("param_oi_tidx",
 593 :               &amp;rstan::stan_fit&lt;stan_model,
 594 :                                boost::random::ecuyer1988&gt;::param_oi_tidx)
 595 :       .method("grad_log_prob",
 596 :               &amp;rstan::stan_fit&lt;stan_model,
 597 :                                boost::random::ecuyer1988&gt;::grad_log_prob)
 598 :       .method("log_prob",
 599 :               &amp;rstan::stan_fit&lt;stan_model, boost::random::ecuyer1988&gt;::log_prob)
 600 :       .method("unconstrain_pars",
 601 :               &amp;rstan::stan_fit&lt;stan_model,
 602 :                                boost::random::ecuyer1988&gt;::unconstrain_pars)
 603 :       .method("constrain_pars",
 604 :               &amp;rstan::stan_fit&lt;stan_model,
 605 :                                boost::random::ecuyer1988&gt;::constrain_pars)
 606 :       .method(
 607 :           "num_pars_unconstrained",
 608 :           &amp;rstan::stan_fit&lt;stan_model,
 609 :                            boost::random::ecuyer1988&gt;::num_pars_unconstrained)
 610 :       .method(
 611 :           "unconstrained_param_names",
 612 :           &amp;rstan::stan_fit&lt;
 613 :               stan_model, boost::random::ecuyer1988&gt;::unconstrained_param_names)
 614 :       .method(
 615 :           "constrained_param_names",
 616 :           &amp;rstan::stan_fit&lt;stan_model,
 617 :                            boost::random::ecuyer1988&gt;::constrained_param_names)
 618 :       .method("standalone_gqs",
 619 :               &amp;rstan::stan_fit&lt;stan_model,
 620 :                                boost::random::ecuyer1988&gt;::standalone_gqs);
 621 : }
 622 : 
 623 : 
 624 : // declarations
 625 : extern "C" {
 626 : SEXP file5ad445a85e1a( ) ;
 627 : }
 628 : 
 629 : // definition
 630 : SEXP file5ad445a85e1a() {
 631 :  return Rcpp::wrap("anon_model");
 632 : }</code></pre>
</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>GaltonBoth <span class="ot">&lt;-</span> mosaicData<span class="sc">::</span>Galton <span class="sc">%&gt;%</span> </span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">midparent =</span> (father <span class="sc">+</span> mother)<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">group =</span> <span class="fu">as.numeric</span>(<span class="fu">factor</span>(sex)) <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span>   <span class="co"># 0s y 1s</span></span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(family) <span class="sc">%&gt;%</span> </span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">1</span>)</span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>galton2_stanfit <span class="ot">&lt;-</span></span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sampling</span>(</span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a>    modelo_galton2,</span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">N =</span> <span class="fu">nrow</span>(GaltonBoth),</span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> GaltonBoth<span class="sc">$</span>midparent,</span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> GaltonBoth<span class="sc">$</span>height,</span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">g =</span> GaltonBoth<span class="sc">$</span>group        <span class="co"># 0s y 1s</span></span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">chains =</span> <span class="dv">4</span>,</span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">iter =</span> <span class="dv">2000</span>,</span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">warmup =</span> <span class="dv">1000</span></span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a>  )  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 6.8e-05 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0.68 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 11.227 seconds (Warm-up)
Chain 1:                15.205 seconds (Sampling)
Chain 1:                26.432 seconds (Total)
Chain 1: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 2).
Chain 2: 
Chain 2: Gradient evaluation took 0.000159 seconds
Chain 2: 1000 transitions using 10 leapfrog steps per transition would take 1.59 seconds.
Chain 2: Adjust your expectations accordingly!
Chain 2: 
Chain 2: 
Chain 2: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 2: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 2: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 2: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 2: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 2: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 2: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 2: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 2: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 2: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 2: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 2: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 2: 
Chain 2:  Elapsed Time: 15.381 seconds (Warm-up)
Chain 2:                11.521 seconds (Sampling)
Chain 2:                26.902 seconds (Total)
Chain 2: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 3).
Chain 3: 
Chain 3: Gradient evaluation took 0.000132 seconds
Chain 3: 1000 transitions using 10 leapfrog steps per transition would take 1.32 seconds.
Chain 3: Adjust your expectations accordingly!
Chain 3: 
Chain 3: 
Chain 3: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 3: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 3: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 3: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 3: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 3: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 3: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 3: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 3: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 3: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 3: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 3: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 3: 
Chain 3:  Elapsed Time: 13.158 seconds (Warm-up)
Chain 3:                9.654 seconds (Sampling)
Chain 3:                22.812 seconds (Total)
Chain 3: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 4).
Chain 4: 
Chain 4: Gradient evaluation took 5.6e-05 seconds
Chain 4: 1000 transitions using 10 leapfrog steps per transition would take 0.56 seconds.
Chain 4: Adjust your expectations accordingly!
Chain 4: 
Chain 4: 
Chain 4: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 4: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 4: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 4: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 4: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 4: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 4: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 4: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 4: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 4: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 4: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 4: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 4: 
Chain 4:  Elapsed Time: 11.622 seconds (Warm-up)
Chain 4:                12.732 seconds (Sampling)
Chain 4:                24.354 seconds (Total)
Chain 4: </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>galton2_stanfit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Stan model: anon_model.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

              mean se_mean    sd    2.5%     25%     50%     75%   97.5% n_eff
beta0        18.00    0.14  5.80    6.80   14.24   17.85   21.84   29.44  1712
beta1         0.69    0.00  0.09    0.52    0.63    0.69    0.75    0.86  1716
beta2         4.84    0.01  0.31    4.24    4.64    4.84    5.05    5.47  2762
sigma         2.17    0.00  0.12    1.95    2.08    2.16    2.24    2.41  2567
nuMinusOne   48.37    0.62 31.38   10.70   25.33   40.47   63.36  131.25  2564
log10nu       1.61    0.01  0.27    1.07    1.42    1.62    1.81    2.12  2188
nu           49.37    0.62 31.38   11.70   26.33   41.47   64.36  132.25  2564
lp__       -321.22    0.04  1.56 -324.97 -321.99 -320.89 -320.08 -319.19  1543
           Rhat
beta0         1
beta1         1
beta2         1
sigma         1
nuMinusOne    1
log10nu       1
nu            1
lp__          1

Samples were drawn using NUTS(diag_e) at Mon Jun 17 22:12:06 2024.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>galton2_mcmc <span class="ot">&lt;-</span> <span class="fu">as.mcmc.list</span>(galton2_stanfit)</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>Post_galton2 <span class="ot">&lt;-</span> <span class="fu">posterior</span>(galton2_stanfit)</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(galton2_mcmc, <span class="at">regex_pars =</span> <span class="fu">c</span>(<span class="st">"beta"</span>, <span class="st">"sigma"</span>, <span class="st">"log10nu"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_post</span>(Post_galton2<span class="sc">$</span>beta2, <span class="at">xlab =</span> <span class="st">"beta2"</span>, <span class="at">hdi_prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-35-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$posterior
         ESS     mean   median    mode
var1 2689.49 4.842483 4.839496 4.78823

$hdi
  prob       lo       hi
1 0.95 4.253709 5.470268</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(galton2_mcmc, <span class="at">regex_pars =</span> <span class="st">"beta2"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_areas</span>(galton2_mcmc, <span class="at">regex_pars =</span> <span class="st">"log10nu"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-36-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GaltonBoth, <span class="dv">3</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 8
# Groups:   family [3]
  family father mother sex   height nkids midparent group
  &lt;fct&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;  &lt;dbl&gt; &lt;int&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 1        78.5   67   F       69.2     4      72.8     0
2 10       74     65.5 F       65.5     1      69.8     0
3 100      69     66   M       71       3      67.5     1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>yind <span class="ot">&lt;-</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">posterior_calc</span>(</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>    galton2_stanfit, </span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>    yind <span class="sc">~</span> beta0 <span class="sc">+</span> beta1 <span class="sc">*</span> midparent <span class="sc">+</span> beta2 <span class="sc">*</span> group <span class="sc">+</span> </span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a>           <span class="fu">rt</span>(<span class="fu">nrow</span>(GaltonBoth), <span class="at">df =</span> nu) <span class="sc">*</span> sigma,</span>
<span id="cb153-6"><a href="#cb153-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> GaltonBoth</span>
<span id="cb153-7"><a href="#cb153-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb153-8"><a href="#cb153-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ppc_intervals_grouped</span>(</span>
<span id="cb153-9"><a href="#cb153-9" aria-hidden="true" tabindex="-1"></a>  GaltonBoth<span class="sc">$</span>height, yind, <span class="at">group =</span> GaltonBoth<span class="sc">$</span>sex, </span>
<span id="cb153-10"><a href="#cb153-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> GaltonBoth<span class="sc">$</span>midparent) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch17-galton2-ppc-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>ppc_data <span class="ot">&lt;-</span></span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ppc_intervals_data</span>(</span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a>  GaltonBoth<span class="sc">$</span>height, yind, <span class="at">group =</span> GaltonBoth<span class="sc">$</span>sex, </span>
<span id="cb154-4"><a href="#cb154-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> GaltonBoth<span class="sc">$</span>midparent)</span>
<span id="cb154-5"><a href="#cb154-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb154-6"><a href="#cb154-6" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(ppc_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 197
Columns: 11
$ y_id        &lt;int&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17,…
$ y_obs       &lt;dbl&gt; 69.2, 65.5, 71.0, 68.5, 68.5, 61.0, 64.5, 68.5, 65.5, 65.5…
$ group       &lt;fct&gt; F, F, M, M, M, F, F, F, F, F, F, F, M, M, F, M, F, M, F, F…
$ x           &lt;dbl&gt; 72.75, 69.75, 67.50, 67.85, 67.50, 67.75, 68.00, 67.75, 67…
$ outer_width &lt;dbl&gt; 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9, 0.9…
$ inner_width &lt;dbl&gt; 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5…
$ ll          &lt;dbl&gt; 64.65196, 62.68861, 65.99708, 66.06591, 65.82074, 61.33906…
$ l           &lt;dbl&gt; 66.83026, 64.79340, 68.09751, 68.28307, 68.08200, 63.43360…
$ m           &lt;dbl&gt; 68.37563, 66.25828, 69.60212, 69.80352, 69.59443, 64.90686…
$ h           &lt;dbl&gt; 69.98837, 67.80330, 71.07960, 71.33653, 71.10201, 66.42513…
$ hh          &lt;dbl&gt; 72.26733, 69.97070, 73.30426, 73.58624, 73.37297, 68.64609…</code></pre>
</div>
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_ribbon</span>(ll <span class="sc">+</span> hh <span class="sc">~</span> x, <span class="at">fill =</span> <span class="sc">~</span> group, <span class="at">data =</span> ppc_data) <span class="sc">%&gt;%</span></span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_ribbon</span>(l <span class="sc">+</span> h <span class="sc">~</span> x, <span class="at">fill =</span> <span class="sc">~</span> group, <span class="at">data =</span> ppc_data) <span class="sc">%&gt;%</span></span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_point</span>(y_obs <span class="sc">~</span> x, <span class="at">color =</span>  <span class="sc">~</span> group, <span class="at">data =</span> ppc_data) <span class="sc">%&gt;%</span></span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_facet_grid</span>(group <span class="sc">~</span> .)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch17-galton2-ppc-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Entonces, ¿qué aprendemos de todos los resultados anteriores?</p>
<ul>
<li>Los diagnósticos sugieren que el modelo está convergiendo adecuadamente.</li>
<li>Las comprobaciones de predicción posterior no muestran discrepancias importantes entre el modelo y los datos. (Por lo tanto, nuestra restricción de que las pendientes de las líneas sean iguales para hombres y mujeres parece estar bien).</li>
<li>La “distribución del ruido” parece estar bien aproximada con una distribución normal (La mayoría de la distribución posterior para <span class="math inline">\(\log_{10}(\nu)\)</span> está por encima de 1.5.)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(galton2_stanfit), <span class="at">regex_pars =</span> <span class="st">"nu"</span>, <span class="at">prob =</span> <span class="fl">0.90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         par       lo        hi      mode prob
1 nuMinusOne 7.761967 91.333268 22.897250  0.9
2    log10nu 1.157328  2.051937  1.562134  0.9
3         nu 8.761967 92.333268 23.897250  0.9</code></pre>
</div>
</div>
<ul>
<li>La nueva característica de este modelo es <span class="math inline">\(\beta_2\)</span>, que cuantifica la diferencia en las alturas <em>promedio</em> de hombres y mujeres <em>cuyos padres tienen las mismas alturas</em>. Aquí está el HDI del 95% para <span class="math inline">\(\beta_2\)</span> (junto con la pendiente y el intercepto):</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(galton2_stanfit), <span class="at">regex_pars =</span> <span class="st">"beta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    par        lo         hi       mode prob
1 beta0 7.4417524 29.9278400 14.6711459 0.95
2 beta1 0.5277311  0.8643556  0.7419232 0.95
3 beta2 4.2537094  5.4702680  4.7873137 0.95</code></pre>
</div>
</div>
</section>
</section>
<section id="regresión-lineal-múltiple" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="regresión-lineal-múltiple"><span class="header-section-number">6.3</span> Regresión Lineal Múltiple</h2>
<section id="ejemplo-sat" class="level3" data-number="6.3.1">
<h3 data-number="6.3.1" class="anchored" data-anchor-id="ejemplo-sat"><span class="header-section-number">6.3.1</span> Ejemplo: SAT</h3>
<p>¿Gastar más en educación resulta en puntajes SAT más altos? La prueba SAT es un examen estandarizado utilizado para la admisión a universidades en los Estados Unidos, que evalúa habilidades en lectura crítica, matemáticas y escritura. Datos de 1999 pueden usarse para explorar esta pregunta. Entre otras cosas, los datos incluyen el puntaje SAT promedio total (en una escala de 400 a 1600) y la cantidad de dinero gastado en educación (en miles de dólares por estudiante) en cada estado de Estados Unidos.</p>
<p>Como primer intento, podríamos ajustar un modelo lineal (sat vs expend). Usando centrado, el núcleo del modelo se ve así:</p>
<pre><code>  for (i in 1:length(y)) {
    y[i]   ~ dt(mu[i], 1/sigma^2, nu)
    mu[i] &lt;- alpha0 + alpha1 * (x[i] - mean(x))
  }</code></pre>
<p><code>alpha1</code> mide cuánto mejora el rendimiento en el SAT por cada $1000 gastados en educación en un estado. Para ajustar el modelo, necesitamos priors para nuestros cuatro parámetros:</p>
<ul>
<li><code>nu</code>: Podemos usar nuestro exponencial desplazado habitual.</li>
<li><code>sigma</code>: {Unif}(?, ?)</li>
<li><code>alpha0</code>: {Norm}(?, ?)</li>
<li><code>alpha1</code>: {Norm}(0, ?)</li>
</ul>
<p>Las interrogaciones dependen de la escala de nuestras covariables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>sat_model <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(y)) {</span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>    y[i]   <span class="sc">~</span> <span class="fu">dt</span>(mu[i], <span class="dv">1</span><span class="sc">/</span>sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a>    mu[i] <span class="ot">&lt;-</span> alpha0 <span class="sc">+</span> alpha1 <span class="sc">*</span> (x[i] <span class="sc">-</span> <span class="fu">mean</span>(x))</span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="fl">29.0</span>)</span>
<span id="cb162-7"><a href="#cb162-7" aria-hidden="true" tabindex="-1"></a>  nu        <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb162-8"><a href="#cb162-8" aria-hidden="true" tabindex="-1"></a>  alpha0     <span class="sc">~</span> <span class="fu">dnorm</span>(alpha0mean, <span class="dv">1</span> <span class="sc">/</span> alpha0sd<span class="sc">^</span><span class="dv">2</span>) </span>
<span id="cb162-9"><a href="#cb162-9" aria-hidden="true" tabindex="-1"></a>  alpha1     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha1sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb162-10"><a href="#cb162-10" aria-hidden="true" tabindex="-1"></a>  sigma      <span class="sc">~</span> <span class="fu">dunif</span>(sigma_lo, sigma_hi <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb162-11"><a href="#cb162-11" aria-hidden="true" tabindex="-1"></a>  log10nu   <span class="ot">&lt;-</span> <span class="fu">log</span>(nu) <span class="sc">/</span> <span class="fu">log</span>(<span class="dv">10</span>)    <span class="co"># log10(nu)</span></span>
<span id="cb162-12"><a href="#cb162-12" aria-hidden="true" tabindex="-1"></a>  beta0     <span class="ot">&lt;-</span> alpha0 <span class="sc">-</span> <span class="fu">mean</span>(x) <span class="sc">*</span> alpha1          <span class="co"># intercepto verdadero</span></span>
<span id="cb162-13"><a href="#cb162-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Entonces, ¿cómo llenamos los signos de interrogación para este conjunto de datos?</p>
<ul>
<li><p><code>sigma</code>: {Unif}(?,?)</p>
<p>Esto cuantifica la cantidad de variación de un estado a otro entre estados que tienen el mismo gasto por estudiante. La escala del SAT varía de 400 a 1600. Los promedios estatales no estarán cerca de los extremos de esta escala. Una ventana de 6 órdenes de magnitud alrededor de 1 da <strong>{</strong>Unif}(0.001, 1000), ambos extremos de los cuales están bastante lejos de lo que creemos razonable.</p></li>
<li><p><code>alpha0</code>: {Norm}(?, ?)</p>
<p><code>alpha0</code> mide el puntaje SAT promedio para los estados que gastan una cantidad promedio. Dado que los SAT promedio están alrededor de 1000, algo como <strong>{</strong>Norm}(1000, 100) parece razonable.</p></li>
<li><p><code>alpha1</code>: {Norm}(0, ?)</p>
<p>Este es el más complicado. La pendiente de una línea de regresión no puede ser mucho más que <span class="math inline">\(\frac{SD_y}{SD_x}\)</span>, por lo que podemos estimar esa relación o calcularla a partir de nuestros datos para guiar nuestra elección de prior.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>sat_jags <span class="ot">&lt;-</span> </span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sat_model,</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> SAT<span class="sc">$</span>sat,</span>
<span id="cb163-6"><a href="#cb163-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> SAT<span class="sc">$</span>expend,</span>
<span id="cb163-7"><a href="#cb163-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0mean =</span> <span class="dv">1000</span>,    <span class="co"># SAT scores are roughly 500 + 500</span></span>
<span id="cb163-8"><a href="#cb163-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0sd   =</span> <span class="dv">100</span>,     <span class="co"># broad prior on scale of 400 - 1600</span></span>
<span id="cb163-9"><a href="#cb163-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha1sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>expend),</span>
<span id="cb163-10"><a href="#cb163-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_lo =</span> <span class="fl">0.001</span>,     <span class="co"># 3 o.m. less than 1</span></span>
<span id="cb163-11"><a href="#cb163-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_hi =</span> <span class="dv">1000</span>       <span class="co"># 3 o.m. greater than 1</span></span>
<span id="cb163-12"><a href="#cb163-12" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb163-13"><a href="#cb163-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"nu"</span>, <span class="st">"log10nu"</span>, <span class="st">"alpha0"</span>, <span class="st">"beta0"</span>, <span class="st">"alpha1"</span>, <span class="st">"sigma"</span>),</span>
<span id="cb163-14"><a href="#cb163-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter   =</span> <span class="dv">4000</span>,</span>
<span id="cb163-15"><a href="#cb163-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.burnin =</span> <span class="dv">1000</span>,</span>
<span id="cb163-16"><a href="#cb163-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span></span>
<span id="cb163-17"><a href="#cb163-17" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>sat_jags</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Bugs model at "/tmp/Rtmp0d30cr/model5ad4696cd1e5.txt", fit using jags,
 3 chains, each with 4000 iterations (first 1000 discarded), n.thin = 3
 n.sims = 3000 iterations saved
          mu.vect sd.vect     2.5%      25%      50%      75%    97.5%  Rhat
alpha0    966.242  10.209  946.097  959.521  966.020  972.944  986.428 1.001
alpha1    -21.549   7.545  -36.480  -26.344  -21.595  -16.572   -6.669 1.001
beta0    1093.493  45.991 1002.139 1062.673 1092.818 1124.048 1185.625 1.002
log10nu     1.503   0.325    0.833    1.281    1.518    1.732    2.088 1.003
nu         41.285  30.531    6.807   19.087   32.990   53.980  122.544 1.003
sigma      70.124   7.998   56.089   64.395   69.584   75.396   87.097 1.001
deviance  568.593   2.710  565.317  566.633  567.943  569.892  575.463 1.001
         n.eff
alpha0    3000
alpha1    2100
beta0     1900
log10nu    980
nu         770
sigma     3000
deviance  3000

For each parameter, n.eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor (at convergence, Rhat=1).

DIC info (using the rule, pD = var(deviance)/2)
pD = 3.7 and DIC = 572.3
DIC is an estimate of expected predictive error (lower deviance is better).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag_mcmc</span>(<span class="fu">as.mcmc</span>(sat_jags))</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc</span>(sat_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-42-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Nuestro interés principal es <code>alpha1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary_df</span>(sat_jags) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(param <span class="sc">==</span> <span class="st">"alpha1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 10
  param   mean    sd `2.5%` `25%` `50%` `75%` `97.5%`  Rhat n.eff
  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 alpha1 -21.5  7.55  -36.5 -26.3 -21.6 -16.6   -6.67  1.00  2100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_post</span>(<span class="fu">posterior</span>(sat_jags)<span class="sc">$</span>alpha1, <span class="at">xlab =</span> <span class="st">"alpha1"</span>, <span class="at">ROPE =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$posterior
          ESS      mean    median      mode
var1 3247.167 -21.54868 -21.59495 -21.59789

$hdi
  prob        lo        hi
1 0.95 -36.55332 -6.671171

$ROPE
  lo hi P(&lt; ROPE) P(in ROPE)    P(&gt; ROPE)
1 -5  5 0.9863333      0.013 0.0006666667</code></pre>
</div>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(sat_jags), <span class="at">pars =</span> <span class="st">"alpha1"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     par        lo        hi      mode prob
1 alpha1 -36.55332 -6.671171 -22.86641 0.95</code></pre>
</div>
</div>
<p>Esto parece extraño: ¿Realmente podemos aumentar los puntajes SAT reduciendo la financiación a las escuelas? Quizás deberíamos observar los datos en bruto con nuestro modelo superpuesto.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(sat <span class="sc">~</span> expend, <span class="at">data =</span> SAT) <span class="sc">%&gt;%</span></span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gf_abline</span>(<span class="at">slope =</span> <span class="sc">~</span> alpha1, <span class="at">intercept =</span> <span class="sc">~</span> beta0, </span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">data =</span> <span class="fu">posterior</span>(sat_jags) <span class="sc">%&gt;%</span> <span class="fu">sample_n</span>(<span class="dv">2000</span>),</span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">alpha =</span> <span class="fl">0.01</span>, <span class="at">color =</span> <span class="st">"steelblue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Hay mucha dispersión, y la tendencia negativa está fuertemente influenciada por los 4 estados que gastan más (y tienen puntajes SAT relativamente bajos). Para solventar este problema vamos a incluir más covariables.</p>
<p>Tenemos algunos datos adicionales sobre cada estado. Vamos a ajustar un modelo con dos predictores: <code>expend</code> y <code>frac</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>SAT <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     state expend ratio salary frac verbal math  sat
1  Alabama  4.405  17.2 31.144    8    491  538 1029
2   Alaska  8.963  17.6 47.951   47    445  489  934
3  Arizona  4.778  19.3 32.175   27    448  496  944
4 Arkansas  4.459  17.1 28.934    6    482  523 1005</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>sat_model2 <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(y)) {</span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>    y[i]   <span class="sc">~</span> <span class="fu">dt</span>(mu[i], <span class="dv">1</span><span class="sc">/</span>sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a>    mu[i] <span class="ot">&lt;-</span> alpha0 <span class="sc">+</span> alpha1 <span class="sc">*</span> (x1[i] <span class="sc">-</span> <span class="fu">mean</span>(x1)) <span class="sc">+</span> alpha2 <span class="sc">*</span> (x2[i] <span class="sc">-</span> <span class="fu">mean</span>(x2))</span>
<span id="cb176-5"><a href="#cb176-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb176-6"><a href="#cb176-6" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="fl">29.0</span>)</span>
<span id="cb176-7"><a href="#cb176-7" aria-hidden="true" tabindex="-1"></a>  nu        <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb176-8"><a href="#cb176-8" aria-hidden="true" tabindex="-1"></a>  alpha0     <span class="sc">~</span> <span class="fu">dnorm</span>(alpha0mean, <span class="dv">1</span> <span class="sc">/</span> alpha0sd<span class="sc">^</span><span class="dv">2</span>) </span>
<span id="cb176-9"><a href="#cb176-9" aria-hidden="true" tabindex="-1"></a>  alpha1     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha1sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb176-10"><a href="#cb176-10" aria-hidden="true" tabindex="-1"></a>  alpha2     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha2sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb176-11"><a href="#cb176-11" aria-hidden="true" tabindex="-1"></a>  sigma      <span class="sc">~</span> <span class="fu">dunif</span>(sigma_lo, sigma_hi <span class="sc">*</span> <span class="dv">1000</span>)</span>
<span id="cb176-12"><a href="#cb176-12" aria-hidden="true" tabindex="-1"></a>  beta0     <span class="ot">&lt;-</span> alpha0 <span class="sc">-</span> <span class="fu">mean</span>(x1) <span class="sc">*</span> alpha1 <span class="sc">-</span> <span class="fu">mean</span>(x2) <span class="sc">*</span> alpha2</span>
<span id="cb176-13"><a href="#cb176-13" aria-hidden="true" tabindex="-1"></a>  log10nu   <span class="ot">&lt;-</span> <span class="fu">log</span>(nu) <span class="sc">/</span> <span class="fu">log</span>(<span class="dv">10</span>)</span>
<span id="cb176-14"><a href="#cb176-14" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>sat2_jags <span class="ot">&lt;-</span> </span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sat_model2,</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> SAT<span class="sc">$</span>sat,</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">x1 =</span> SAT<span class="sc">$</span>expend,</span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">x2 =</span> SAT<span class="sc">$</span>frac,</span>
<span id="cb177-8"><a href="#cb177-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0mean =</span> <span class="dv">1000</span>,    <span class="co"># Los puntajes SAT son aproximadamente 500 + 500</span></span>
<span id="cb177-9"><a href="#cb177-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0sd   =</span> <span class="dv">100</span>,     <span class="co"># Prior amplio en la escala de 400 - 1600</span></span>
<span id="cb177-10"><a href="#cb177-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha1sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>expend),</span>
<span id="cb177-11"><a href="#cb177-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha2sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>frac),</span>
<span id="cb177-12"><a href="#cb177-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_lo =</span> <span class="fl">0.001</span>,</span>
<span id="cb177-13"><a href="#cb177-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_hi =</span> <span class="dv">1000</span></span>
<span id="cb177-14"><a href="#cb177-14" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb177-15"><a href="#cb177-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"log10nu"</span>, <span class="st">"alpha0"</span>, <span class="st">"alpha1"</span>, <span class="st">"alpha2"</span>, <span class="st">"beta0"</span>,<span class="st">"sigma"</span>),</span>
<span id="cb177-16"><a href="#cb177-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter   =</span> <span class="dv">4000</span>,</span>
<span id="cb177-17"><a href="#cb177-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.burnin =</span> <span class="dv">1000</span>,</span>
<span id="cb177-18"><a href="#cb177-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span></span>
<span id="cb177-19"><a href="#cb177-19" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>sat2_jags</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Bugs model at "/tmp/Rtmp0d30cr/model5ad4630ec062.txt", fit using jags,
 3 chains, each with 4000 iterations (first 1000 discarded), n.thin = 3
 n.sims = 3000 iterations saved
         mu.vect sd.vect    2.5%     25%     50%      75%    97.5%  Rhat n.eff
alpha0   965.920   4.659 956.865 962.728 966.026  969.149  974.758 1.001  2100
alpha1    12.862   4.509   3.927   9.869  12.874   15.858   21.955 1.001  3000
alpha2    -2.885   0.234  -3.347  -3.043  -2.883   -2.729   -2.440 1.001  3000
beta0    991.621  22.846 946.274 976.400 991.427 1006.406 1037.458 1.001  3000
log10nu    1.367   0.370   0.633   1.107   1.380    1.637    2.039 1.001  3000
sigma     31.488   3.918  24.249  28.834  31.256   33.890   39.656 1.001  3000
deviance 491.177   3.088 487.322 488.929 490.501  492.711  498.730 1.006   780

For each parameter, n.eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor (at convergence, Rhat=1).

DIC info (using the rule, pD = var(deviance)/2)
pD = 4.8 and DIC = 495.9
DIC is an estimate of expected predictive error (lower deviance is better).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag_mcmc</span>(<span class="fu">as.mcmc</span>(sat2_jags))</span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc</span>(sat2_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch18-sat2-jags-look-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch18-sat2-jags-look-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary_df</span>(sat2_jags) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(param <span class="sc">==</span> <span class="st">"alpha1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 10
  param   mean    sd `2.5%` `25%` `50%` `75%` `97.5%`  Rhat n.eff
  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 alpha1  12.9  4.51   3.93  9.87  12.9  15.9    22.0  1.00  3000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_post</span>(<span class="fu">posterior</span>(sat2_jags)<span class="sc">$</span>alpha1, <span class="at">xlab =</span> <span class="st">"alpha1"</span>, <span class="at">ROPE =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch18-sat2-alpha1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$posterior
          ESS     mean   median     mode
var1 2490.794 12.86232 12.87438 12.99886

$hdi
  prob       lo      hi
1 0.95 3.447806 21.3269

$ROPE
  lo hi P(&lt; ROPE) P(in ROPE) P(&gt; ROPE)
1 -5  5         0 0.03933333 0.9606667</code></pre>
</div>
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(sat2_jags), <span class="at">pars =</span> <span class="st">"alpha1"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     par       lo      hi     mode prob
1 alpha1 3.447806 21.3269 11.07646 0.95</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary_df</span>(sat2_jags) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(param <span class="sc">==</span> <span class="st">"alpha2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 10
  param   mean    sd `2.5%` `25%` `50%` `75%` `97.5%`  Rhat n.eff
  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 alpha2 -2.88 0.234  -3.35 -3.04 -2.88 -2.73   -2.44  1.00  3000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_post</span>(<span class="fu">posterior</span>(sat2_jags)<span class="sc">$</span>alpha2, <span class="at">xlab =</span> <span class="st">"alpha2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/ch18-sat2-alpha2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$posterior
          ESS      mean    median      mode
var1 2897.033 -2.884702 -2.883404 -2.823664

$hdi
  prob        lo        hi
1 0.95 -3.334813 -2.435674</code></pre>
</div>
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hdi</span>(<span class="fu">posterior</span>(sat2_jags), <span class="at">pars =</span> <span class="st">"alpha2"</span>, <span class="at">prob =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     par        lo        hi      mode prob
1 alpha2 -3.334813 -2.435674 -2.895194 0.95</code></pre>
</div>
</div>
<p>Las covariables parecen estar correlacionadas:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb193"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gf_point</span>(expend <span class="sc">~</span> frac, <span class="at">data =</span> SAT) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>por lo que un modelo con interacción puede ser más apropiado:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb194"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a>sat_model3 <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb194-2"><a href="#cb194-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(y)) {</span>
<span id="cb194-3"><a href="#cb194-3" aria-hidden="true" tabindex="-1"></a>    y[i]   <span class="sc">~</span> <span class="fu">dt</span>(mu[i], <span class="dv">1</span><span class="sc">/</span>sigma<span class="sc">^</span><span class="dv">2</span>, nu)</span>
<span id="cb194-4"><a href="#cb194-4" aria-hidden="true" tabindex="-1"></a>    mu[i] <span class="ot">&lt;-</span> alpha0 <span class="sc">+</span> alpha1 <span class="sc">*</span> (x1[i] <span class="sc">-</span> <span class="fu">mean</span>(x1)) <span class="sc">+</span> alpha2 <span class="sc">*</span> (x2[i] <span class="sc">-</span> <span class="fu">mean</span>(x2)) <span class="sc">+</span>  alpha3 <span class="sc">*</span> (x3[i] <span class="sc">-</span> <span class="fu">mean</span>(x3))</span>
<span id="cb194-5"><a href="#cb194-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb194-6"><a href="#cb194-6" aria-hidden="true" tabindex="-1"></a>  nuMinusOne <span class="sc">~</span> <span class="fu">dexp</span>(<span class="dv">1</span><span class="sc">/</span><span class="fl">29.0</span>)</span>
<span id="cb194-7"><a href="#cb194-7" aria-hidden="true" tabindex="-1"></a>  nu        <span class="ot">&lt;-</span> nuMinusOne <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb194-8"><a href="#cb194-8" aria-hidden="true" tabindex="-1"></a>  alpha0     <span class="sc">~</span> <span class="fu">dnorm</span>(alpha0mean, <span class="dv">1</span> <span class="sc">/</span> alpha0sd<span class="sc">^</span><span class="dv">2</span>) </span>
<span id="cb194-9"><a href="#cb194-9" aria-hidden="true" tabindex="-1"></a>  alpha1     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha1sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb194-10"><a href="#cb194-10" aria-hidden="true" tabindex="-1"></a>  alpha2     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha2sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb194-11"><a href="#cb194-11" aria-hidden="true" tabindex="-1"></a>  alpha3     <span class="sc">~</span> <span class="fu">dnorm</span>(<span class="dv">0</span>, <span class="dv">1</span> <span class="sc">/</span> alpha3sd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb194-12"><a href="#cb194-12" aria-hidden="true" tabindex="-1"></a>  sigma      <span class="sc">~</span> <span class="fu">dunif</span>(sigma_lo, sigma_hi)</span>
<span id="cb194-13"><a href="#cb194-13" aria-hidden="true" tabindex="-1"></a>  beta0     <span class="ot">&lt;-</span> alpha0 <span class="sc">-</span> <span class="fu">mean</span>(x1) <span class="sc">*</span> alpha1 <span class="sc">-</span> <span class="fu">mean</span>(x2) <span class="sc">*</span> alpha2</span>
<span id="cb194-14"><a href="#cb194-14" aria-hidden="true" tabindex="-1"></a>  log10nu   <span class="ot">&lt;-</span> <span class="fu">log</span>(nu) <span class="sc">/</span> <span class="fu">log</span>(<span class="dv">10</span>)</span>
<span id="cb194-15"><a href="#cb194-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a>sat3_jags <span class="ot">&lt;-</span> </span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jags</span>(</span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">model =</span> sat_model3,</span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">list</span>(</span>
<span id="cb195-5"><a href="#cb195-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> SAT<span class="sc">$</span>sat,</span>
<span id="cb195-6"><a href="#cb195-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">x1 =</span> SAT<span class="sc">$</span>expend,</span>
<span id="cb195-7"><a href="#cb195-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">x2 =</span> SAT<span class="sc">$</span>frac,</span>
<span id="cb195-8"><a href="#cb195-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">x3 =</span> SAT<span class="sc">$</span>frac <span class="sc">*</span> SAT<span class="sc">$</span>expend,</span>
<span id="cb195-9"><a href="#cb195-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0mean =</span> <span class="dv">1000</span>,    <span class="co"># SAT scores are roughly 500 + 500</span></span>
<span id="cb195-10"><a href="#cb195-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha0sd   =</span> <span class="dv">100</span>,     <span class="co"># broad prior on scale of 400 - 1600</span></span>
<span id="cb195-11"><a href="#cb195-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha1sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>expend),</span>
<span id="cb195-12"><a href="#cb195-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha2sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>frac),</span>
<span id="cb195-13"><a href="#cb195-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">alpha3sd   =</span> <span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>frac <span class="sc">*</span> SAT<span class="sc">$</span>expend),</span>
<span id="cb195-14"><a href="#cb195-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_lo =</span> <span class="fl">0.001</span>,</span>
<span id="cb195-15"><a href="#cb195-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">sigma_hi =</span> <span class="dv">1000</span></span>
<span id="cb195-16"><a href="#cb195-16" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb195-17"><a href="#cb195-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">parameters.to.save =</span> <span class="fu">c</span>(<span class="st">"log10nu"</span>, <span class="st">"alpha0"</span>, <span class="st">"alpha1"</span>, <span class="st">"alpha2"</span>, <span class="st">"alpha3"</span>, <span class="st">"beta0"</span>,<span class="st">"sigma"</span>),</span>
<span id="cb195-18"><a href="#cb195-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.iter   =</span> <span class="dv">20000</span>,</span>
<span id="cb195-19"><a href="#cb195-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.burnin =</span> <span class="dv">1000</span>,</span>
<span id="cb195-20"><a href="#cb195-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span></span>
<span id="cb195-21"><a href="#cb195-21" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb196"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a>sat3_jags</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Bugs model at "/tmp/Rtmp0d30cr/model5ad417b7508f.txt", fit using jags,
 3 chains, each with 20000 iterations (first 1000 discarded), n.thin = 19
 n.sims = 3000 iterations saved
          mu.vect sd.vect    2.5%      25%      50%      75%    97.5%  Rhat
alpha0    965.935   4.635 956.969  962.886  966.006  968.999  975.006 1.001
alpha1      2.270   8.183 -13.759   -3.221    2.319    7.792   17.822 1.001
alpha2     -4.110   0.836  -5.700   -4.683   -4.115   -3.557   -2.456 1.002
alpha3      0.212   0.139  -0.062    0.119    0.211    0.306    0.474 1.002
beta0    1097.368  72.897 955.671 1048.895 1097.161 1147.277 1236.971 1.001
log10nu     1.405   0.361   0.665    1.159    1.420    1.661    2.062 1.002
sigma      31.178   3.853  24.235   28.502   30.941   33.590   39.466 1.001
deviance  489.174   3.464 484.614  486.657  488.469  490.996  497.872 1.002
         n.eff
alpha0    3000
alpha1    3000
alpha2    1800
alpha3    1700
beta0     3000
log10nu   1700
sigma     3000
deviance  1900

For each parameter, n.eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor (at convergence, Rhat=1).

DIC info (using the rule, pD = var(deviance)/2)
pD = 6.0 and DIC = 495.2
DIC is an estimate of expected predictive error (lower deviance is better).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag_mcmc</span>(<span class="fu">as.mcmc</span>(sat3_jags))</span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc</span>(sat3_jags))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-51-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_pairs</span>(<span class="fu">as.mcmc</span>(sat3_jags), <span class="at">regex_pars =</span> <span class="st">"alpha"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-51-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="ajuste-de-un-modelo-lineal-con-brms" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="ajuste-de-un-modelo-lineal-con-brms"><span class="header-section-number">6.4</span> Ajuste de un modelo lineal con brms</h2>
<p>El paquete brms proporciona una forma simplificada de describir modelos lineales generalizados y ajustarlos con Stan. La función <code>brm()</code> convierte una descripción concisa del modelo en código Stan, lo compila y lo ejecuta. Aquí hay un modelo lineal con <code>sat</code> como respuesta, y <code>expend</code>, <code>frac</code>, y una interacción como predictores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(brms)  </span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>sat3_brm <span class="ot">&lt;-</span> <span class="fu">brm</span>(sat <span class="sc">~</span> expend <span class="sc">*</span> frac, <span class="at">data =</span> SAT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Compiling Stan program...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Start sampling</code></pre>
</div>
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a>sat3_stan <span class="ot">&lt;-</span> <span class="fu">stanfit</span>(sat3_brm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Stan maneja mejor los parámetros correlacionados que JAGS.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a>sat3_stan</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Stan model: anon_model.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

                 mean se_mean    sd    2.5%     25%     50%     75%   97.5%
b_Intercept   1057.07    1.14 42.99  973.33 1028.43 1057.12 1086.24 1142.70
b_expend         0.59    0.21  8.02  -15.09   -4.92    0.62    6.03   16.24
b_frac          -4.23    0.02  0.84   -5.91   -4.78   -4.22   -3.67   -2.61
b_expend:frac    0.24    0.00  0.14   -0.03    0.14    0.24    0.33    0.51
sigma           32.50    0.07  3.36   26.78   30.14   32.19   34.52   39.92
Intercept      965.78    0.09  4.75  956.61  962.63  965.74  968.94  975.06
lprior         -10.32    0.00  0.03  -10.38  -10.34  -10.32  -10.30  -10.28
lp__          -251.30    0.04  1.68 -255.53 -252.15 -250.92 -250.08 -249.11
              n_eff Rhat
b_Intercept    1432    1
b_expend       1492    1
b_frac         1288    1
b_expend:frac  1240    1
sigma          2192    1
Intercept      2802    1
lprior         2509    1
lp__           1593    1

Samples were drawn using NUTS(diag_e) at Mon Jun 17 22:15:45 2024.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc.list</span>(sat3_stan))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Podemos usar <code>stancode()</code> para extraer el código Stan utilizado para ajustar el modelo.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb207"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stancode</span>(sat3_brm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>// generated with brms 2.21.0
functions {
}
data {
  int&lt;lower=1&gt; N;  // total number of observations
  vector[N] Y;  // response variable
  int&lt;lower=1&gt; K;  // number of population-level effects
  matrix[N, K] X;  // population-level design matrix
  int&lt;lower=1&gt; Kc;  // number of population-level effects after centering
  int prior_only;  // should the likelihood be ignored?
}
transformed data {
  matrix[N, Kc] Xc;  // centered version of X without an intercept
  vector[Kc] means_X;  // column means of X before centering
  for (i in 2:K) {
    means_X[i - 1] = mean(X[, i]);
    Xc[, i - 1] = X[, i] - means_X[i - 1];
  }
}
parameters {
  vector[Kc] b;  // regression coefficients
  real Intercept;  // temporary intercept for centered predictors
  real&lt;lower=0&gt; sigma;  // dispersion parameter
}
transformed parameters {
  real lprior = 0;  // prior contributions to the log posterior
  lprior += student_t_lpdf(Intercept | 3, 945.5, 84.5);
  lprior += student_t_lpdf(sigma | 3, 0, 84.5)
    - 1 * student_t_lccdf(0 | 3, 0, 84.5);
}
model {
  // likelihood including constants
  if (!prior_only) {
    target += normal_id_glm_lpdf(Y | Xc, Intercept, b, sigma);
  }
  // priors including constants
  target += lprior;
}
generated quantities {
  // actual population-level intercept
  real b_Intercept = Intercept - dot_product(means_X, b);
}</code></pre>
</div>
</div>
<p>Podemos usar <code>standata()</code> para mostrar los datos que <code>brm()</code> pasa a Stan.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">standata</span>(sat3_brm) <span class="sc">%&gt;%</span> </span>
<span id="cb209-2"><a href="#cb209-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(head)  <span class="co"># trunca la salida para ahorrar espacio</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$N
[1] 50

$Y
[1] 1029  934  944 1005  902  980

$K
[1] 4

$Kc
[1] 3

$X
  Intercept expend frac expend:frac
1         1  4.405    8      35.240
2         1  8.963   47     421.261
3         1  4.778   27     129.006
4         1  4.459    6      26.754
5         1  4.992   45     224.640
6         1  5.443   29     157.847

$prior_only
[1] 0</code></pre>
</div>
</div>
<p>Supongamos que queremos construir un modelo que tenga el mismo prior y verosimilitud que nuestro modelo JAGS. Aquí están algunos valores que necesitaremos.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb211"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>expend)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 219.6073</code></pre>
</div>
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>frac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 11.18293</code></pre>
</div>
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> <span class="sc">*</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>sat) <span class="sc">/</span> <span class="fu">sd</span>(SAT<span class="sc">$</span>frac <span class="sc">*</span> SAT<span class="sc">$</span>expend)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.45156</code></pre>
</div>
</div>
<p>Para usar una distribución t para la respuesta, usamos <code>family = student()</code>. Para establecer los priors, es útil saber cuáles serán los nombres de los parámetros y cuáles serían los priors predeterminados si no hacemos nada. (Si no se lista ningún prior, se usará un prior plano impropio.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_prior</span>(</span>
<span id="cb217-2"><a href="#cb217-2" aria-hidden="true" tabindex="-1"></a>  sat <span class="sc">~</span> expend <span class="sc">*</span> frac, <span class="at">data =</span> SAT,</span>
<span id="cb217-3"><a href="#cb217-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">student</span>()   <span class="co"># distribución para la variable de respuesta</span></span>
<span id="cb217-4"><a href="#cb217-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     prior     class        coef group resp dpar nlpar lb ub
                    (flat)         b                                        
                    (flat)         b      expend                            
                    (flat)         b expend:frac                            
                    (flat)         b        frac                            
 student_t(3, 945.5, 84.5) Intercept                                        
             gamma(2, 0.1)        nu                                    1   
     student_t(3, 0, 84.5)     sigma                                    0   
       source
      default
 (vectorized)
 (vectorized)
 (vectorized)
      default
      default
      default</code></pre>
</div>
</div>
<p>Podemos comunicar los priors a <code>brm()</code> de la siguiente manera:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb219"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb219-1"><a href="#cb219-1" aria-hidden="true" tabindex="-1"></a>sat3a_brm <span class="ot">&lt;-</span> </span>
<span id="cb219-2"><a href="#cb219-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">brm</span>(</span>
<span id="cb219-3"><a href="#cb219-3" aria-hidden="true" tabindex="-1"></a>    sat <span class="sc">~</span> expend <span class="sc">*</span> frac, <span class="at">data =</span> SAT,</span>
<span id="cb219-4"><a href="#cb219-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">family =</span> <span class="fu">student</span>(),</span>
<span id="cb219-5"><a href="#cb219-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">prior =</span> <span class="fu">c</span>(</span>
<span id="cb219-6"><a href="#cb219-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"normal(0,220)"</span>, <span class="at">coef =</span> <span class="st">"expend"</span>),</span>
<span id="cb219-7"><a href="#cb219-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"normal(0,11)"</span>, <span class="at">coef =</span> <span class="st">"frac"</span>),</span>
<span id="cb219-8"><a href="#cb219-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"normal(0,1.5)"</span>, <span class="at">coef =</span> <span class="st">"expend:frac"</span>),</span>
<span id="cb219-9"><a href="#cb219-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"normal(1000, 100)"</span>, <span class="at">class =</span> <span class="st">"Intercept"</span>),</span>
<span id="cb219-10"><a href="#cb219-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"exponential(1/30.0)"</span>, <span class="at">class =</span> <span class="st">"nu"</span>),</span>
<span id="cb219-11"><a href="#cb219-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set_prior</span>(<span class="st">"uniform(0.001,1000)"</span>, <span class="at">class =</span> <span class="st">"sigma"</span>)</span>
<span id="cb219-12"><a href="#cb219-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb219-13"><a href="#cb219-13" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: It appears as if you have specified an upper bounded prior on a parameter that has no natural upper bound.
If this is really what you want, please specify argument 'ub' of 'set_prior' appropriately.
Warning occurred for prior 
&lt;lower=0&gt; sigma ~ uniform(0.001,1000)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Compiling Stan program...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Start sampling</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 1).
Chain 1: 
Chain 1: Gradient evaluation took 1.7e-05 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 0.17 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 1: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 1: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 1: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 1: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 1: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 1: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 1: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 1: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 1: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 1: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 1: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 1: 
Chain 1:  Elapsed Time: 0.36 seconds (Warm-up)
Chain 1:                0.216 seconds (Sampling)
Chain 1:                0.576 seconds (Total)
Chain 1: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 2).
Chain 2: 
Chain 2: Gradient evaluation took 1.1e-05 seconds
Chain 2: 1000 transitions using 10 leapfrog steps per transition would take 0.11 seconds.
Chain 2: Adjust your expectations accordingly!
Chain 2: 
Chain 2: 
Chain 2: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 2: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 2: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 2: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 2: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 2: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 2: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 2: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 2: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 2: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 2: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 2: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 2: 
Chain 2:  Elapsed Time: 0.314 seconds (Warm-up)
Chain 2:                0.237 seconds (Sampling)
Chain 2:                0.551 seconds (Total)
Chain 2: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 3).
Chain 3: 
Chain 3: Gradient evaluation took 1.1e-05 seconds
Chain 3: 1000 transitions using 10 leapfrog steps per transition would take 0.11 seconds.
Chain 3: Adjust your expectations accordingly!
Chain 3: 
Chain 3: 
Chain 3: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 3: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 3: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 3: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 3: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 3: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 3: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 3: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 3: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 3: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 3: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 3: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 3: 
Chain 3:  Elapsed Time: 0.296 seconds (Warm-up)
Chain 3:                0.203 seconds (Sampling)
Chain 3:                0.499 seconds (Total)
Chain 3: 

SAMPLING FOR MODEL 'anon_model' NOW (CHAIN 4).
Chain 4: 
Chain 4: Gradient evaluation took 1.2e-05 seconds
Chain 4: 1000 transitions using 10 leapfrog steps per transition would take 0.12 seconds.
Chain 4: Adjust your expectations accordingly!
Chain 4: 
Chain 4: 
Chain 4: Iteration:    1 / 2000 [  0%]  (Warmup)
Chain 4: Iteration:  200 / 2000 [ 10%]  (Warmup)
Chain 4: Iteration:  400 / 2000 [ 20%]  (Warmup)
Chain 4: Iteration:  600 / 2000 [ 30%]  (Warmup)
Chain 4: Iteration:  800 / 2000 [ 40%]  (Warmup)
Chain 4: Iteration: 1000 / 2000 [ 50%]  (Warmup)
Chain 4: Iteration: 1001 / 2000 [ 50%]  (Sampling)
Chain 4: Iteration: 1200 / 2000 [ 60%]  (Sampling)
Chain 4: Iteration: 1400 / 2000 [ 70%]  (Sampling)
Chain 4: Iteration: 1600 / 2000 [ 80%]  (Sampling)
Chain 4: Iteration: 1800 / 2000 [ 90%]  (Sampling)
Chain 4: Iteration: 2000 / 2000 [100%]  (Sampling)
Chain 4: 
Chain 4:  Elapsed Time: 0.27 seconds (Warm-up)
Chain 4:                0.201 seconds (Sampling)
Chain 4:                0.471 seconds (Total)
Chain 4: </code></pre>
</div>
<div class="sourceCode cell-code" id="cb224"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" aria-hidden="true" tabindex="-1"></a>sat3a_stan <span class="ot">&lt;-</span> <span class="fu">stanfit</span>(sat3a_brm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb225"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb225-1"><a href="#cb225-1" aria-hidden="true" tabindex="-1"></a>sat3a_stan</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Inference for Stan model: anon_model.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

                 mean se_mean    sd    2.5%     25%     50%     75%   97.5%
b_Intercept   1050.14    1.06 43.42  961.46 1020.46 1050.15 1080.20 1133.73
b_expend         2.02    0.20  8.12  -13.72   -3.55    1.90    7.68   18.31
b_frac          -4.16    0.02  0.85   -5.77   -4.75   -4.16   -3.56   -2.51
b_expend:frac    0.22    0.00  0.14   -0.06    0.12    0.22    0.32    0.49
sigma           31.13    0.08  3.84   24.14   28.55   30.92   33.45   39.14
nu              34.61    0.55 30.22    4.51   13.55   25.16   46.20  117.32
Intercept      965.85    0.08  4.64  957.06  962.74  965.90  968.93  975.13
lprior         -28.06    0.02  1.01  -30.86  -28.44  -27.74  -27.35  -27.04
lp__          -266.12    0.05  1.81 -270.61 -267.10 -265.80 -264.78 -263.60
              n_eff Rhat
b_Intercept    1684    1
b_expend       1719    1
b_frac         1553    1
b_expend:frac  1463    1
sigma          2144    1
nu             3038    1
Intercept      3158    1
lprior         2997    1
lp__           1361    1

Samples were drawn using NUTS(diag_e) at Mon Jun 17 22:19:38 2024.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb227"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb227-1"><a href="#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mcmc_combo</span>(<span class="fu">as.mcmc.list</span>(sat3a_stan))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="estimacion_modelos2_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./estimacion_modelos1.html" class="pagination-link" aria-label="Estimación de Modelos Bayesianos (Parte 1)">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Estimación de Modelos Bayesianos (Parte 1)</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>